<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 9 More Examples | ComplexHeatmap Complete Reference</title>
  <meta name="description" content="Complex heatmaps are efficient to visualize associations between different sources of data sets and reveal potential patterns. Here the ComplexHeatmap R package provides a highly flexible way to arrange multiple heatmaps and supports various annotation graphics. This book is the complete reference to ComplexHeatmap pacakge." />
  <meta name="generator" content="bookdown 0.14 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 9 More Examples | ComplexHeatmap Complete Reference" />
  <meta property="og:type" content="book" />
  <meta property="og:url" content="https://jokergoo.github.io/ComplexHeatmap-reference/book" />
  <meta property="og:image" content="https://jokergoo.github.io/ComplexHeatmap-reference/bookcomplexheatmap-cover.jpg" />
  <meta property="og:description" content="Complex heatmaps are efficient to visualize associations between different sources of data sets and reveal potential patterns. Here the ComplexHeatmap R package provides a highly flexible way to arrange multiple heatmaps and supports various annotation graphics. This book is the complete reference to ComplexHeatmap pacakge." />
  <meta name="github-repo" content="jokergoo/ComplexHeatmap-reference" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 9 More Examples | ComplexHeatmap Complete Reference" />
  
  <meta name="twitter:description" content="Complex heatmaps are efficient to visualize associations between different sources of data sets and reveal potential patterns. Here the ComplexHeatmap R package provides a highly flexible way to arrange multiple heatmaps and supports various annotation graphics. This book is the complete reference to ComplexHeatmap pacakge." />
  <meta name="twitter:image" content="https://jokergoo.github.io/ComplexHeatmap-reference/bookcomplexheatmap-cover.jpg" />

<meta name="author" content="Zuguang Gu" />



  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="upset-plot.html"/>
<link rel="next" href="other-high-level-plots.html"/>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />











<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; } /* Keyword */
code > span.dt { color: #902000; } /* DataType */
code > span.dv { color: #40a070; } /* DecVal */
code > span.bn { color: #40a070; } /* BaseN */
code > span.fl { color: #40a070; } /* Float */
code > span.ch { color: #4070a0; } /* Char */
code > span.st { color: #4070a0; } /* String */
code > span.co { color: #60a0b0; font-style: italic; } /* Comment */
code > span.ot { color: #007020; } /* Other */
code > span.al { color: #ff0000; font-weight: bold; } /* Alert */
code > span.fu { color: #06287e; } /* Function */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #880000; } /* Constant */
code > span.sc { color: #4070a0; } /* SpecialChar */
code > span.vs { color: #4070a0; } /* VerbatimString */
code > span.ss { color: #bb6688; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #19177c; } /* Variable */
code > span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code > span.op { color: #666666; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #bc7a00; } /* Preprocessor */
code > span.at { color: #7d9029; } /* Attribute */
code > span.do { color: #ba2121; font-style: italic; } /* Documentation */
code > span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">ComplexHeatmap Complete Reference</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>About</a></li>
<li class="chapter" data-level="1" data-path="introduction.html"><a href="introduction.html"><i class="fa fa-check"></i><b>1</b> Introduction</a><ul>
<li class="chapter" data-level="1.1" data-path="introduction.html"><a href="introduction.html#general-design"><i class="fa fa-check"></i><b>1.1</b> General design</a></li>
<li class="chapter" data-level="1.2" data-path="introduction.html"><a href="introduction.html#a-brief-description-of-following-chapters"><i class="fa fa-check"></i><b>1.2</b> A brief description of following chapters</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="a-single-heatmap.html"><a href="a-single-heatmap.html"><i class="fa fa-check"></i><b>2</b> A Single Heatmap</a><ul>
<li class="chapter" data-level="2.1" data-path="a-single-heatmap.html"><a href="a-single-heatmap.html#colors"><i class="fa fa-check"></i><b>2.1</b> Colors</a></li>
<li class="chapter" data-level="2.2" data-path="a-single-heatmap.html"><a href="a-single-heatmap.html#heatmap-titles"><i class="fa fa-check"></i><b>2.2</b> Titles</a></li>
<li class="chapter" data-level="2.3" data-path="a-single-heatmap.html"><a href="a-single-heatmap.html#clustering"><i class="fa fa-check"></i><b>2.3</b> Clustering</a><ul>
<li class="chapter" data-level="2.3.1" data-path="a-single-heatmap.html"><a href="a-single-heatmap.html#distance-methods"><i class="fa fa-check"></i><b>2.3.1</b> Distance methods</a></li>
<li class="chapter" data-level="2.3.2" data-path="a-single-heatmap.html"><a href="a-single-heatmap.html#clustering-methods"><i class="fa fa-check"></i><b>2.3.2</b> Clustering methods</a></li>
<li class="chapter" data-level="2.3.3" data-path="a-single-heatmap.html"><a href="a-single-heatmap.html#render-dendrograms"><i class="fa fa-check"></i><b>2.3.3</b> Render dendrograms</a></li>
<li class="chapter" data-level="2.3.4" data-path="a-single-heatmap.html"><a href="a-single-heatmap.html#reorder-dendrograms"><i class="fa fa-check"></i><b>2.3.4</b> Reorder dendrograms</a></li>
</ul></li>
<li class="chapter" data-level="2.4" data-path="a-single-heatmap.html"><a href="a-single-heatmap.html#row-and_column_orders"><i class="fa fa-check"></i><b>2.4</b> Set row and column orders</a></li>
<li class="chapter" data-level="2.5" data-path="a-single-heatmap.html"><a href="a-single-heatmap.html#heatmap-seriation"><i class="fa fa-check"></i><b>2.5</b> Seriation</a></li>
<li class="chapter" data-level="2.6" data-path="a-single-heatmap.html"><a href="a-single-heatmap.html#dimension-names"><i class="fa fa-check"></i><b>2.6</b> Dimension names</a></li>
<li class="chapter" data-level="2.7" data-path="a-single-heatmap.html"><a href="a-single-heatmap.html#heatmap-split"><i class="fa fa-check"></i><b>2.7</b> Heatmap split</a><ul>
<li class="chapter" data-level="2.7.1" data-path="a-single-heatmap.html"><a href="a-single-heatmap.html#split-by-kmeans-clustering"><i class="fa fa-check"></i><b>2.7.1</b> Split by k-means clustering</a></li>
<li class="chapter" data-level="2.7.2" data-path="a-single-heatmap.html"><a href="a-single-heatmap.html#split-by-categorical-variables"><i class="fa fa-check"></i><b>2.7.2</b> Split by categorical variables</a></li>
<li class="chapter" data-level="2.7.3" data-path="a-single-heatmap.html"><a href="a-single-heatmap.html#spilt-by-dendrogram"><i class="fa fa-check"></i><b>2.7.3</b> Split by dendrogram</a></li>
<li class="chapter" data-level="2.7.4" data-path="a-single-heatmap.html"><a href="a-single-heatmap.html#order-of-slices"><i class="fa fa-check"></i><b>2.7.4</b> Order of slices</a></li>
<li class="chapter" data-level="2.7.5" data-path="a-single-heatmap.html"><a href="a-single-heatmap.html#titles-for-splitting"><i class="fa fa-check"></i><b>2.7.5</b> Titles for splitting</a></li>
<li class="chapter" data-level="2.7.6" data-path="a-single-heatmap.html"><a href="a-single-heatmap.html#graphic-parameters-for-splitting"><i class="fa fa-check"></i><b>2.7.6</b> Graphic parameters for splitting</a></li>
<li class="chapter" data-level="2.7.7" data-path="a-single-heatmap.html"><a href="a-single-heatmap.html#gaps-between-slices"><i class="fa fa-check"></i><b>2.7.7</b> Gaps between slices</a></li>
<li class="chapter" data-level="2.7.8" data-path="a-single-heatmap.html"><a href="a-single-heatmap.html#split-heatmap-annotations"><i class="fa fa-check"></i><b>2.7.8</b> Split heatmap annotations</a></li>
</ul></li>
<li class="chapter" data-level="2.8" data-path="a-single-heatmap.html"><a href="a-single-heatmap.html#heatmap-as-raster-image"><i class="fa fa-check"></i><b>2.8</b> Heatmap as raster image</a></li>
<li class="chapter" data-level="2.9" data-path="a-single-heatmap.html"><a href="a-single-heatmap.html#customize-the-heatmap-body"><i class="fa fa-check"></i><b>2.9</b> Customize the heatmap body</a><ul>
<li class="chapter" data-level="2.9.1" data-path="a-single-heatmap.html"><a href="a-single-heatmap.html#cell-fun"><i class="fa fa-check"></i><b>2.9.1</b> cell_fun</a></li>
<li class="chapter" data-level="2.9.2" data-path="a-single-heatmap.html"><a href="a-single-heatmap.html#layer-fun"><i class="fa fa-check"></i><b>2.9.2</b> layer_fun</a></li>
</ul></li>
<li class="chapter" data-level="2.10" data-path="a-single-heatmap.html"><a href="a-single-heatmap.html#size-of-the-heatmap"><i class="fa fa-check"></i><b>2.10</b> Size of the heatmap</a></li>
<li class="chapter" data-level="2.11" data-path="a-single-heatmap.html"><a href="a-single-heatmap.html#plot-the-heatmap"><i class="fa fa-check"></i><b>2.11</b> Plot the heatmap</a></li>
<li class="chapter" data-level="2.12" data-path="a-single-heatmap.html"><a href="a-single-heatmap.html#get-orders-and-dendrograms-from-heatmap"><i class="fa fa-check"></i><b>2.12</b> Get orders and dendrograms</a></li>
<li class="chapter" data-level="2.13" data-path="a-single-heatmap.html"><a href="a-single-heatmap.html#subset-a-heatmap"><i class="fa fa-check"></i><b>2.13</b> Subset a heatmap</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="heatmap-annotations.html"><a href="heatmap-annotations.html"><i class="fa fa-check"></i><b>3</b> Heatmap Annotations</a><ul>
<li class="chapter" data-level="3.1" data-path="heatmap-annotations.html"><a href="heatmap-annotations.html#simple-annotation"><i class="fa fa-check"></i><b>3.1</b> Simple annotation</a></li>
<li class="chapter" data-level="3.2" data-path="heatmap-annotations.html"><a href="heatmap-annotations.html#simple-annotation-as-an-annotation-function"><i class="fa fa-check"></i><b>3.2</b> Simple annotation as an annotation function</a></li>
<li class="chapter" data-level="3.3" data-path="heatmap-annotations.html"><a href="heatmap-annotations.html#empty-annotation"><i class="fa fa-check"></i><b>3.3</b> Empty annotation</a></li>
<li class="chapter" data-level="3.4" data-path="heatmap-annotations.html"><a href="heatmap-annotations.html#block-annotation"><i class="fa fa-check"></i><b>3.4</b> Block annotation</a></li>
<li class="chapter" data-level="3.5" data-path="heatmap-annotations.html"><a href="heatmap-annotations.html#image-annotation"><i class="fa fa-check"></i><b>3.5</b> Image annotation</a></li>
<li class="chapter" data-level="3.6" data-path="heatmap-annotations.html"><a href="heatmap-annotations.html#points-annotation"><i class="fa fa-check"></i><b>3.6</b> Points annotation</a></li>
<li class="chapter" data-level="3.7" data-path="heatmap-annotations.html"><a href="heatmap-annotations.html#lines-annotation"><i class="fa fa-check"></i><b>3.7</b> Lines annotation</a></li>
<li class="chapter" data-level="3.8" data-path="heatmap-annotations.html"><a href="heatmap-annotations.html#barplot_annotation"><i class="fa fa-check"></i><b>3.8</b> Barplot annotation</a></li>
<li class="chapter" data-level="3.9" data-path="heatmap-annotations.html"><a href="heatmap-annotations.html#box-annotation"><i class="fa fa-check"></i><b>3.9</b> Boxplot annotation</a></li>
<li class="chapter" data-level="3.10" data-path="heatmap-annotations.html"><a href="heatmap-annotations.html#histogram-annotation"><i class="fa fa-check"></i><b>3.10</b> Histogram annotation</a></li>
<li class="chapter" data-level="3.11" data-path="heatmap-annotations.html"><a href="heatmap-annotations.html#density-annotation"><i class="fa fa-check"></i><b>3.11</b> Density annotation</a></li>
<li class="chapter" data-level="3.12" data-path="heatmap-annotations.html"><a href="heatmap-annotations.html#joyplot-annotation"><i class="fa fa-check"></i><b>3.12</b> Joyplot annotation</a></li>
<li class="chapter" data-level="3.13" data-path="heatmap-annotations.html"><a href="heatmap-annotations.html#horizon-chart-annotation"><i class="fa fa-check"></i><b>3.13</b> Horizon chart annotation</a></li>
<li class="chapter" data-level="3.14" data-path="heatmap-annotations.html"><a href="heatmap-annotations.html#text-annotation"><i class="fa fa-check"></i><b>3.14</b> Text annotation</a></li>
<li class="chapter" data-level="3.15" data-path="heatmap-annotations.html"><a href="heatmap-annotations.html#mark-annotation"><i class="fa fa-check"></i><b>3.15</b> Mark annotation</a></li>
<li class="chapter" data-level="3.16" data-path="heatmap-annotations.html"><a href="heatmap-annotations.html#summary-annotation"><i class="fa fa-check"></i><b>3.16</b> Summary annotation</a></li>
<li class="chapter" data-level="3.17" data-path="heatmap-annotations.html"><a href="heatmap-annotations.html#zoom-annotation"><i class="fa fa-check"></i><b>3.17</b> Zoom annotation</a></li>
<li class="chapter" data-level="3.18" data-path="heatmap-annotations.html"><a href="heatmap-annotations.html#multiple-annotations"><i class="fa fa-check"></i><b>3.18</b> Multiple annotations</a><ul>
<li class="chapter" data-level="3.18.1" data-path="heatmap-annotations.html"><a href="heatmap-annotations.html#size-of-annotations"><i class="fa fa-check"></i><b>3.18.1</b> Size of annotations</a></li>
</ul></li>
<li class="chapter" data-level="3.19" data-path="heatmap-annotations.html"><a href="heatmap-annotations.html#heatmap-annotation-utility-function"><i class="fa fa-check"></i><b>3.19</b> Utility functions</a></li>
<li class="chapter" data-level="3.20" data-path="heatmap-annotations.html"><a href="heatmap-annotations.html#implement-new-annotation-functions"><i class="fa fa-check"></i><b>3.20</b> Implement new annotation functions</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="a-list-of-heatmaps.html"><a href="a-list-of-heatmaps.html"><i class="fa fa-check"></i><b>4</b> A List of Heatmaps</a><ul>
<li class="chapter" data-level="4.1" data-path="a-list-of-heatmaps.html"><a href="a-list-of-heatmaps.html#heatmap-list-titles"><i class="fa fa-check"></i><b>4.1</b> Titles</a></li>
<li class="chapter" data-level="4.2" data-path="a-list-of-heatmaps.html"><a href="a-list-of-heatmaps.html#size-of-heatmaps"><i class="fa fa-check"></i><b>4.2</b> Size of heatmaps</a></li>
<li class="chapter" data-level="4.3" data-path="a-list-of-heatmaps.html"><a href="a-list-of-heatmaps.html#gap-between-heatmaps"><i class="fa fa-check"></i><b>4.3</b> Gap between heatmaps</a></li>
<li class="chapter" data-level="4.4" data-path="a-list-of-heatmaps.html"><a href="a-list-of-heatmaps.html#automatic-adjustment-to-the-main-heatmap"><i class="fa fa-check"></i><b>4.4</b> Automatic adjustment to the main heatmap</a></li>
<li class="chapter" data-level="4.5" data-path="a-list-of-heatmaps.html"><a href="a-list-of-heatmaps.html#control-main-heatmap-in-draw-function"><i class="fa fa-check"></i><b>4.5</b> Control main heatmap in draw() function</a></li>
<li class="chapter" data-level="4.6" data-path="a-list-of-heatmaps.html"><a href="a-list-of-heatmaps.html#annotations-as-components-are-adjusted"><i class="fa fa-check"></i><b>4.6</b> Annotations as components are adjusted</a></li>
<li class="chapter" data-level="4.7" data-path="a-list-of-heatmaps.html"><a href="a-list-of-heatmaps.html#concatenate-with-annotations"><i class="fa fa-check"></i><b>4.7</b> Concatenate with annotations</a></li>
<li class="chapter" data-level="4.8" data-path="a-list-of-heatmaps.html"><a href="a-list-of-heatmaps.html#concatenate-only-the-annotations"><i class="fa fa-check"></i><b>4.8</b> Concatenate only the annotations</a></li>
<li class="chapter" data-level="4.9" data-path="a-list-of-heatmaps.html"><a href="a-list-of-heatmaps.html#vertical-concatenation"><i class="fa fa-check"></i><b>4.9</b> Vertical concatenation</a></li>
<li class="chapter" data-level="4.10" data-path="a-list-of-heatmaps.html"><a href="a-list-of-heatmaps.html#subset-heatmap-list"><i class="fa fa-check"></i><b>4.10</b> Subset the heatmap list</a></li>
<li class="chapter" data-level="4.11" data-path="a-list-of-heatmaps.html"><a href="a-list-of-heatmaps.html#plot-the-heamtap-list"><i class="fa fa-check"></i><b>4.11</b> Plot the heatmap list</a></li>
<li class="chapter" data-level="4.12" data-path="a-list-of-heatmaps.html"><a href="a-list-of-heatmaps.html#get-orders-and-dendrograms-from-a-list-of-heatmaps"><i class="fa fa-check"></i><b>4.12</b> Get orders and dendrograms</a></li>
<li class="chapter" data-level="4.13" data-path="a-list-of-heatmaps.html"><a href="a-list-of-heatmaps.html#change-parameters-globally"><i class="fa fa-check"></i><b>4.13</b> Change parameters globally</a></li>
<li class="chapter" data-level="4.14" data-path="a-list-of-heatmaps.html"><a href="a-list-of-heatmaps.html#adjust-blank-space-caused-by-annotations"><i class="fa fa-check"></i><b>4.14</b> Adjust blank space caused by annotations</a></li>
<li class="chapter" data-level="4.15" data-path="a-list-of-heatmaps.html"><a href="a-list-of-heatmaps.html#manually-increase-space-around-the-plot"><i class="fa fa-check"></i><b>4.15</b> Manually increase space around the plot</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="legends.html"><a href="legends.html"><i class="fa fa-check"></i><b>5</b> Legends</a><ul>
<li class="chapter" data-level="5.1" data-path="legends.html"><a href="legends.html#continuous-legends"><i class="fa fa-check"></i><b>5.1</b> Continuous legends</a></li>
<li class="chapter" data-level="5.2" data-path="legends.html"><a href="legends.html#discrete-legends"><i class="fa fa-check"></i><b>5.2</b> Discrete legends</a></li>
<li class="chapter" data-level="5.3" data-path="legends.html"><a href="legends.html#a-list-of-legends"><i class="fa fa-check"></i><b>5.3</b> A list of legends</a></li>
<li class="chapter" data-level="5.4" data-path="legends.html"><a href="legends.html#heatmap-and-annotation-legends"><i class="fa fa-check"></i><b>5.4</b> Heatmap and annotation legends</a></li>
<li class="chapter" data-level="5.5" data-path="legends.html"><a href="legends.html#add-customized-legends"><i class="fa fa-check"></i><b>5.5</b> Add customized legends</a></li>
<li class="chapter" data-level="5.6" data-path="legends.html"><a href="legends.html#the-side-of-legends"><i class="fa fa-check"></i><b>5.6</b> The side of legends</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="heatmap-decoration.html"><a href="heatmap-decoration.html"><i class="fa fa-check"></i><b>6</b> Heatmap Decoration</a><ul>
<li class="chapter" data-level="6.1" data-path="heatmap-decoration.html"><a href="heatmap-decoration.html#decoration-functions"><i class="fa fa-check"></i><b>6.1</b> Decoration functions</a></li>
<li class="chapter" data-level="6.2" data-path="heatmap-decoration.html"><a href="heatmap-decoration.html#decoration-examples"><i class="fa fa-check"></i><b>6.2</b> Examples</a><ul>
<li class="chapter" data-level="6.2.1" data-path="heatmap-decoration.html"><a href="heatmap-decoration.html#barplot-for-single-column-heatmap"><i class="fa fa-check"></i><b>6.2.1</b> Barplot for single-column heatmap</a></li>
<li class="chapter" data-level="6.2.2" data-path="heatmap-decoration.html"><a href="heatmap-decoration.html#add-titles-for-row-annotations"><i class="fa fa-check"></i><b>6.2.2</b> Add titles for row annotations</a></li>
<li class="chapter" data-level="6.2.3" data-path="heatmap-decoration.html"><a href="heatmap-decoration.html#other-possible-use-of-decorations"><i class="fa fa-check"></i><b>6.2.3</b> Other possible use of decorations</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="7" data-path="oncoprint.html"><a href="oncoprint.html"><i class="fa fa-check"></i><b>7</b> OncoPrint</a><ul>
<li class="chapter" data-level="7.1" data-path="oncoprint.html"><a href="oncoprint.html#oncoprint-general-settings"><i class="fa fa-check"></i><b>7.1</b> General settings</a><ul>
<li class="chapter" data-level="7.1.1" data-path="oncoprint.html"><a href="oncoprint.html#input-data-format"><i class="fa fa-check"></i><b>7.1.1</b> Input data format</a></li>
<li class="chapter" data-level="7.1.2" data-path="oncoprint.html"><a href="oncoprint.html#define-the-alter-fun"><i class="fa fa-check"></i><b>7.1.2</b> Define the alter_fun()</a></li>
<li class="chapter" data-level="7.1.3" data-path="oncoprint.html"><a href="oncoprint.html#oncoprint-background"><i class="fa fa-check"></i><b>7.1.3</b> Background</a></li>
<li class="chapter" data-level="7.1.4" data-path="oncoprint.html"><a href="oncoprint.html#complex-alteration-types"><i class="fa fa-check"></i><b>7.1.4</b> Complex alteration types</a></li>
<li class="chapter" data-level="7.1.5" data-path="oncoprint.html"><a href="oncoprint.html#other-heatmap-related-settings"><i class="fa fa-check"></i><b>7.1.5</b> Other heatmap-related settings</a></li>
</ul></li>
<li class="chapter" data-level="7.2" data-path="oncoprint.html"><a href="oncoprint.html#apply-to-cbioportal-dataset"><i class="fa fa-check"></i><b>7.2</b> Apply to cBioPortal dataset</a><ul>
<li class="chapter" data-level="7.2.1" data-path="oncoprint.html"><a href="oncoprint.html#remove-empty-rows-and-columns"><i class="fa fa-check"></i><b>7.2.1</b> Remove empty rows and columns</a></li>
<li class="chapter" data-level="7.2.2" data-path="oncoprint.html"><a href="oncoprint.html#reorder-the-oncoprint"><i class="fa fa-check"></i><b>7.2.2</b> Reorder the oncoPrint</a></li>
<li class="chapter" data-level="7.2.3" data-path="oncoprint.html"><a href="oncoprint.html#oncoprint-annotations"><i class="fa fa-check"></i><b>7.2.3</b> OncoPrint annotations</a></li>
<li class="chapter" data-level="7.2.4" data-path="oncoprint.html"><a href="oncoprint.html#oncoprint-as-a-heatmap"><i class="fa fa-check"></i><b>7.2.4</b> oncoPrint as a Heatmap</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="8" data-path="upset-plot.html"><a href="upset-plot.html"><i class="fa fa-check"></i><b>8</b> UpSet plot</a><ul>
<li class="chapter" data-level="8.1" data-path="upset-plot.html"><a href="upset-plot.html#input-data"><i class="fa fa-check"></i><b>8.1</b> Input data</a></li>
<li class="chapter" data-level="8.2" data-path="upset-plot.html"><a href="upset-plot.html#upset-mode"><i class="fa fa-check"></i><b>8.2</b> Mode</a></li>
<li class="chapter" data-level="8.3" data-path="upset-plot.html"><a href="upset-plot.html#make-the-combination-matrix"><i class="fa fa-check"></i><b>8.3</b> Make the combination matrix</a></li>
<li class="chapter" data-level="8.4" data-path="upset-plot.html"><a href="upset-plot.html#upset-utility-functions"><i class="fa fa-check"></i><b>8.4</b> Utility functions</a></li>
<li class="chapter" data-level="8.5" data-path="upset-plot.html"><a href="upset-plot.html#upset-making-the-plot"><i class="fa fa-check"></i><b>8.5</b> Make the plot</a></li>
<li class="chapter" data-level="8.6" data-path="upset-plot.html"><a href="upset-plot.html#upset-plots-as-heatmaps"><i class="fa fa-check"></i><b>8.6</b> UpSet plots as heatmaps</a></li>
<li class="chapter" data-level="8.7" data-path="upset-plot.html"><a href="upset-plot.html#example-with-the-movies-dataset"><i class="fa fa-check"></i><b>8.7</b> Example with the movies dataset</a></li>
<li class="chapter" data-level="8.8" data-path="upset-plot.html"><a href="upset-plot.html#example-with-the-genomic-regions"><i class="fa fa-check"></i><b>8.8</b> Example with the genomic regions</a></li>
</ul></li>
<li class="chapter" data-level="9" data-path="more-examples.html"><a href="more-examples.html"><i class="fa fa-check"></i><b>9</b> More Examples</a><ul>
<li class="chapter" data-level="9.1" data-path="more-examples.html"><a href="more-examples.html#add-more-information-for-gene-expression-matrix"><i class="fa fa-check"></i><b>9.1</b> Add more information for gene expression matrix</a></li>
<li class="chapter" data-level="9.2" data-path="more-examples.html"><a href="more-examples.html#the-measles-vaccine-heatmap"><i class="fa fa-check"></i><b>9.2</b> The measles vaccine heatmap</a></li>
<li class="chapter" data-level="9.3" data-path="more-examples.html"><a href="more-examples.html#visualize-cell-heterogeneity-from-single-cell-rnaseq"><i class="fa fa-check"></i><b>9.3</b> Visualize Cell Heterogeneity from Single Cell RNASeq</a></li>
<li class="chapter" data-level="9.4" data-path="more-examples.html"><a href="more-examples.html#correlations-between-methylation-expression-and-other-genomic-features"><i class="fa fa-check"></i><b>9.4</b> Correlations between methylation, expression and other genomic features</a></li>
<li class="chapter" data-level="9.5" data-path="more-examples.html"><a href="more-examples.html#visualize-methylation-profile-with-complex-annotations"><i class="fa fa-check"></i><b>9.5</b> Visualize Methylation Profile with Complex Annotations</a></li>
<li class="chapter" data-level="9.6" data-path="more-examples.html"><a href="more-examples.html#add-multiple-boxplots-for-single-row"><i class="fa fa-check"></i><b>9.6</b> Add multiple boxplots for single row</a></li>
</ul></li>
<li class="chapter" data-level="10" data-path="other-high-level-plots.html"><a href="other-high-level-plots.html"><i class="fa fa-check"></i><b>10</b> Other High-level Plots</a><ul>
<li class="chapter" data-level="10.1" data-path="other-high-level-plots.html"><a href="other-high-level-plots.html#density-heatmap"><i class="fa fa-check"></i><b>10.1</b> Density heatmap</a></li>
<li class="chapter" data-level="10.2" data-path="other-high-level-plots.html"><a href="other-high-level-plots.html#stacked-summary-plot"><i class="fa fa-check"></i><b>10.2</b> Stacked summary plot</a></li>
</ul></li>
<li class="chapter" data-level="11" data-path="other-tricks.html"><a href="other-tricks.html"><i class="fa fa-check"></i><b>11</b> Other Tricks</a><ul>
<li class="chapter" data-level="11.1" data-path="other-tricks.html"><a href="other-tricks.html#set-the-same-cell-size-for-different-heatmaps-with-different-dimensions"><i class="fa fa-check"></i><b>11.1</b> Set the same cell size for different heatmaps with different dimensions</a></li>
</ul></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">ComplexHeatmap Complete Reference</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="more-examples" class="section level1">
<h1><span class="header-section-number">Chapter 9</span> More Examples</h1>
<div id="add-more-information-for-gene-expression-matrix" class="section level2">
<h2><span class="header-section-number">9.1</span> Add more information for gene expression matrix</h2>
<p>Heatmaps are very popular to visualize gene expression matrix. Rows in the matrix correspond to genes and more information on these genes can be attached after the expression heatmap.</p>
<p>In following example, the big heatmap visualizes relative expression for genes (expression for each gene is scaled). On the right we put the absolute expression level of genes as a single-column heatmap. The gene length and gene type (i.e. protein coding or lincRNA) are also put as heatmap annotations or heatmaps.</p>
<p>On the very left of the heatmaps, there are colored rectangles drawn by <code>anno_block()</code> to identify the five clusters from k-means clustering. On top of the “base mean” and “gene type” heatmaps, there are summary plots (barplots and boxplots) showing the statistics or distributions of the data points in the five clusters.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(ComplexHeatmap)
<span class="kw">library</span>(circlize)

expr =<span class="st"> </span><span class="kw">readRDS</span>(<span class="kw">system.file</span>(<span class="dt">package =</span> <span class="st">&quot;ComplexHeatmap&quot;</span>, <span class="st">&quot;extdata&quot;</span>, <span class="st">&quot;gene_expression.rds&quot;</span>))
mat =<span class="st"> </span><span class="kw">as.matrix</span>(expr[, <span class="kw">grep</span>(<span class="st">&quot;cell&quot;</span>, <span class="kw">colnames</span>(expr))])
base_mean =<span class="st"> </span><span class="kw">rowMeans</span>(mat)
mat_scaled =<span class="st"> </span><span class="kw">t</span>(<span class="kw">apply</span>(mat, <span class="dv">1</span>, scale))

type =<span class="st"> </span><span class="kw">gsub</span>(<span class="st">&quot;s</span><span class="ch">\\</span><span class="st">d+_&quot;</span>, <span class="st">&quot;&quot;</span>, <span class="kw">colnames</span>(mat))
ha =<span class="st"> </span><span class="kw">HeatmapAnnotation</span>(<span class="dt">type =</span> type, <span class="dt">annotation_name_side =</span> <span class="st">&quot;left&quot;</span>)

ht_list =<span class="st"> </span><span class="kw">Heatmap</span>(mat_scaled, <span class="dt">name =</span> <span class="st">&quot;expression&quot;</span>, <span class="dt">row_km =</span> <span class="dv">5</span>, 
    <span class="dt">col =</span> <span class="kw">colorRamp2</span>(<span class="kw">c</span>(<span class="op">-</span><span class="dv">2</span>, <span class="dv">0</span>, <span class="dv">2</span>), <span class="kw">c</span>(<span class="st">&quot;green&quot;</span>, <span class="st">&quot;white&quot;</span>, <span class="st">&quot;red&quot;</span>)),
    <span class="dt">top_annotation =</span> ha, 
    <span class="dt">show_column_names =</span> <span class="ot">FALSE</span>, <span class="dt">row_title =</span> <span class="ot">NULL</span>, <span class="dt">show_row_dend =</span> <span class="ot">FALSE</span>) <span class="op">+</span>
<span class="kw">Heatmap</span>(base_mean, <span class="dt">name =</span> <span class="st">&quot;base mean&quot;</span>, 
    <span class="dt">top_annotation =</span> <span class="kw">HeatmapAnnotation</span>(<span class="dt">summary =</span> <span class="kw">anno_summary</span>(<span class="dt">gp =</span> <span class="kw">gpar</span>(<span class="dt">fill =</span> <span class="dv">2</span><span class="op">:</span><span class="dv">6</span>), 
        <span class="dt">height =</span> <span class="kw">unit</span>(<span class="dv">2</span>, <span class="st">&quot;cm&quot;</span>))),
    <span class="dt">width =</span> <span class="kw">unit</span>(<span class="dv">15</span>, <span class="st">&quot;mm&quot;</span>)) <span class="op">+</span>
<span class="kw">rowAnnotation</span>(<span class="dt">length =</span> <span class="kw">anno_points</span>(expr<span class="op">$</span>length, <span class="dt">pch =</span> <span class="dv">16</span>, <span class="dt">size =</span> <span class="kw">unit</span>(<span class="dv">1</span>, <span class="st">&quot;mm&quot;</span>), 
    <span class="dt">axis_param =</span> <span class="kw">list</span>(<span class="dt">at =</span> <span class="kw">c</span>(<span class="dv">0</span>, <span class="fl">2e5</span>, <span class="fl">4e5</span>, <span class="fl">6e5</span>), 
        <span class="dt">labels =</span> <span class="kw">c</span>(<span class="st">&quot;0kb&quot;</span>, <span class="st">&quot;200kb&quot;</span>, <span class="st">&quot;400kb&quot;</span>, <span class="st">&quot;600kb&quot;</span>)),
    <span class="dt">width =</span> <span class="kw">unit</span>(<span class="dv">2</span>, <span class="st">&quot;cm&quot;</span>))) <span class="op">+</span>
<span class="kw">Heatmap</span>(expr<span class="op">$</span>type, <span class="dt">name =</span> <span class="st">&quot;gene type&quot;</span>, 
    <span class="dt">top_annotation =</span> <span class="kw">HeatmapAnnotation</span>(<span class="dt">summary =</span> <span class="kw">anno_summary</span>(<span class="dt">height =</span> <span class="kw">unit</span>(<span class="dv">2</span>, <span class="st">&quot;cm&quot;</span>))),
    <span class="dt">width =</span> <span class="kw">unit</span>(<span class="dv">15</span>, <span class="st">&quot;mm&quot;</span>))

ht_list =<span class="st"> </span><span class="kw">rowAnnotation</span>(<span class="dt">block =</span> <span class="kw">anno_block</span>(<span class="dt">gp =</span> <span class="kw">gpar</span>(<span class="dt">fill =</span> <span class="dv">2</span><span class="op">:</span><span class="dv">6</span>, <span class="dt">col =</span> <span class="ot">NA</span>)), 
    <span class="dt">width =</span> <span class="kw">unit</span>(<span class="dv">2</span>, <span class="st">&quot;mm&quot;</span>)) <span class="op">+</span><span class="st"> </span>ht_list

<span class="kw">draw</span>(ht_list, <span class="dt">ht_gap =</span> <span class="kw">unit</span>(<span class="dv">5</span>, <span class="st">&quot;mm&quot;</span>))</code></pre></div>
<p><img src="09-examples_files/figure-html/expression_example-1.png" width="960" style="display: block; margin: auto;" /></p>
</div>
<div id="the-measles-vaccine-heatmap" class="section level2">
<h2><span class="header-section-number">9.2</span> The measles vaccine heatmap</h2>
<p>Following code reproduces the heatmap introduced <a href="https://biomickwatson.wordpress.com/2015/04/09/recreating-a-famous-visualisation/">here</a> and <a href="https://benjaminlmoore.wordpress.com/2015/04/09/recreating-the-vaccination-heatmaps-in-r/">here</a>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">mat =<span class="st"> </span><span class="kw">readRDS</span>(<span class="kw">system.file</span>(<span class="st">&quot;extdata&quot;</span>, <span class="st">&quot;measles.rds&quot;</span>, <span class="dt">package =</span> <span class="st">&quot;ComplexHeatmap&quot;</span>))
ha1 =<span class="st"> </span><span class="kw">HeatmapAnnotation</span>(
    <span class="dt">dist1 =</span> <span class="kw">anno_barplot</span>(
        <span class="kw">colSums</span>(mat), 
        <span class="dt">bar_width =</span> <span class="dv">1</span>, 
        <span class="dt">gp =</span> <span class="kw">gpar</span>(<span class="dt">col =</span> <span class="st">&quot;white&quot;</span>, <span class="dt">fill =</span> <span class="st">&quot;#FFE200&quot;</span>), 
        <span class="dt">border =</span> <span class="ot">FALSE</span>,
        <span class="dt">axis_param =</span> <span class="kw">list</span>(<span class="dt">at =</span> <span class="kw">c</span>(<span class="dv">0</span>, <span class="fl">2e5</span>, <span class="fl">4e5</span>, <span class="fl">6e5</span>, <span class="fl">8e5</span>),
            <span class="dt">labels =</span> <span class="kw">c</span>(<span class="st">&quot;0&quot;</span>, <span class="st">&quot;200k&quot;</span>, <span class="st">&quot;400k&quot;</span>, <span class="st">&quot;600k&quot;</span>, <span class="st">&quot;800k&quot;</span>)),
        <span class="dt">height =</span> <span class="kw">unit</span>(<span class="dv">2</span>, <span class="st">&quot;cm&quot;</span>)
    ), <span class="dt">show_annotation_name =</span> <span class="ot">FALSE</span>)
ha2 =<span class="st"> </span><span class="kw">rowAnnotation</span>(
    <span class="dt">dist2 =</span> <span class="kw">anno_barplot</span>(
        <span class="kw">rowSums</span>(mat), 
        <span class="dt">bar_width =</span> <span class="dv">1</span>, 
        <span class="dt">gp =</span> <span class="kw">gpar</span>(<span class="dt">col =</span> <span class="st">&quot;white&quot;</span>, <span class="dt">fill =</span> <span class="st">&quot;#FFE200&quot;</span>), 
        <span class="dt">border =</span> <span class="ot">FALSE</span>,
        <span class="dt">axis_param =</span> <span class="kw">list</span>(<span class="dt">at =</span> <span class="kw">c</span>(<span class="dv">0</span>, <span class="fl">5e5</span>, <span class="fl">1e6</span>, <span class="fl">1.5e6</span>),
            <span class="dt">labels =</span> <span class="kw">c</span>(<span class="st">&quot;0&quot;</span>, <span class="st">&quot;500k&quot;</span>, <span class="st">&quot;1m&quot;</span>, <span class="st">&quot;1.5m&quot;</span>)),
        <span class="dt">width =</span> <span class="kw">unit</span>(<span class="dv">2</span>, <span class="st">&quot;cm&quot;</span>)
    ), <span class="dt">show_annotation_name =</span> <span class="ot">FALSE</span>)
year_text =<span class="st"> </span><span class="kw">as.numeric</span>(<span class="kw">colnames</span>(mat))
year_text[year_text <span class="op">%%</span><span class="st"> </span><span class="dv">10</span> <span class="op">!=</span><span class="st"> </span><span class="dv">0</span>] =<span class="st"> &quot;&quot;</span>
ha_column =<span class="st"> </span><span class="kw">HeatmapAnnotation</span>(
    <span class="dt">year =</span> <span class="kw">anno_text</span>(year_text, <span class="dt">rot =</span> <span class="dv">0</span>, <span class="dt">location =</span> <span class="kw">unit</span>(<span class="dv">1</span>, <span class="st">&quot;npc&quot;</span>), <span class="dt">just =</span> <span class="st">&quot;top&quot;</span>)
)
col_fun =<span class="st"> </span><span class="kw">colorRamp2</span>(<span class="kw">c</span>(<span class="dv">0</span>, <span class="dv">800</span>, <span class="dv">1000</span>, <span class="dv">127000</span>), <span class="kw">c</span>(<span class="st">&quot;white&quot;</span>, <span class="st">&quot;cornflowerblue&quot;</span>, <span class="st">&quot;yellow&quot;</span>, <span class="st">&quot;red&quot;</span>))
ht_list =<span class="st"> </span><span class="kw">Heatmap</span>(mat, <span class="dt">name =</span> <span class="st">&quot;cases&quot;</span>, <span class="dt">col =</span> col_fun,
    <span class="dt">cluster_columns =</span> <span class="ot">FALSE</span>, <span class="dt">show_row_dend =</span> <span class="ot">FALSE</span>, <span class="dt">rect_gp =</span> <span class="kw">gpar</span>(<span class="dt">col=</span> <span class="st">&quot;white&quot;</span>), 
    <span class="dt">show_column_names =</span> <span class="ot">FALSE</span>,
    <span class="dt">row_names_side =</span> <span class="st">&quot;left&quot;</span>, <span class="dt">row_names_gp =</span> <span class="kw">gpar</span>(<span class="dt">fontsize =</span> <span class="dv">8</span>),
    <span class="dt">column_title =</span> <span class="st">&#39;Measles cases in US states 1930-2001</span><span class="ch">\n</span><span class="st">Vaccine introduced 1961&#39;</span>,
    <span class="dt">top_annotation =</span> ha1, <span class="dt">bottom_annotation =</span> ha_column,
    <span class="dt">heatmap_legend_param =</span> <span class="kw">list</span>(<span class="dt">at =</span> <span class="kw">c</span>(<span class="dv">0</span>, <span class="fl">5e4</span>, <span class="fl">1e5</span>, <span class="fl">1.5e5</span>), 
        <span class="dt">labels =</span> <span class="kw">c</span>(<span class="st">&quot;0&quot;</span>, <span class="st">&quot;50k&quot;</span>, <span class="st">&quot;100k&quot;</span>, <span class="st">&quot;150k&quot;</span>))) <span class="op">+</span><span class="st"> </span>ha2
<span class="kw">draw</span>(ht_list, <span class="dt">ht_gap =</span> <span class="kw">unit</span>(<span class="dv">3</span>, <span class="st">&quot;mm&quot;</span>))
<span class="kw">decorate_heatmap_body</span>(<span class="st">&quot;cases&quot;</span>, {
    i =<span class="st"> </span><span class="kw">which</span>(<span class="kw">colnames</span>(mat) <span class="op">==</span><span class="st"> &quot;1961&quot;</span>)
    x =<span class="st"> </span>i<span class="op">/</span><span class="kw">ncol</span>(mat)
    <span class="kw">grid.lines</span>(<span class="kw">c</span>(x, x), <span class="kw">c</span>(<span class="dv">0</span>, <span class="dv">1</span>), <span class="dt">gp =</span> <span class="kw">gpar</span>(<span class="dt">lwd =</span> <span class="dv">2</span>, <span class="dt">lty =</span> <span class="dv">2</span>))
    <span class="kw">grid.text</span>(<span class="st">&quot;Vaccine introduced&quot;</span>, x, <span class="kw">unit</span>(<span class="dv">1</span>, <span class="st">&quot;npc&quot;</span>) <span class="op">+</span><span class="st"> </span><span class="kw">unit</span>(<span class="dv">5</span>, <span class="st">&quot;mm&quot;</span>))
})</code></pre></div>
<p><img src="09-examples_files/figure-html/unnamed-chunk-3-1.png" width="960" style="display: block; margin: auto;" /></p>
</div>
<div id="visualize-cell-heterogeneity-from-single-cell-rnaseq" class="section level2">
<h2><span class="header-section-number">9.3</span> Visualize Cell Heterogeneity from Single Cell RNASeq</h2>
<p>In this example, single cell RNA-Seq data for mouse T-cells is visualized to show the heterogeneity of cells. The data (<code>mouse_scRNAseq_corrected.txt</code>) is from <a href="http://www.nature.com/nbt/journal/v33/n2/full/nbt.3102.html">Buettner et al., 2015</a>, supplementary data 1, sheet “<strong>Cell-cycle corrected gene expr</strong>”. You can get <code>mouse_scRNAseq_corrected.txt</code> <a href="https://jokergoo.github.io/ComplexHeatmap-reference/data/mouse_scRNAseq_corrected">here</a>.</p>
<p>In following code, duplicated genes are removed.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">expr =<span class="st"> </span><span class="kw">read.table</span>(<span class="st">&quot;data/mouse_scRNAseq_corrected.txt&quot;</span>, <span class="dt">sep =</span> <span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>, <span class="dt">header =</span> <span class="ot">TRUE</span>)
expr =<span class="st"> </span>expr[<span class="op">!</span><span class="kw">duplicated</span>(expr[[<span class="dv">1</span>]]), ]
<span class="kw">rownames</span>(expr) =<span class="st"> </span>expr[[<span class="dv">1</span>]]
expr =<span class="st"> </span>expr[<span class="op">-</span><span class="dv">1</span>]
expr =<span class="st"> </span><span class="kw">as.matrix</span>(expr)</code></pre></div>
<p>Genes that are not expressed in more than half of the cells are filtered out.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">expr =<span class="st"> </span>expr[<span class="kw">apply</span>(expr, <span class="dv">1</span>, <span class="cf">function</span>(x) <span class="kw">sum</span>(x <span class="op">&gt;</span><span class="st"> </span><span class="dv">0</span>)<span class="op">/</span><span class="kw">length</span>(x) <span class="op">&gt;</span><span class="st"> </span><span class="fl">0.5</span>), , drop =<span class="st"> </span><span class="ot">FALSE</span>]</code></pre></div>
<p>The <code>get_correlated_variable_rows()</code> function is defined here. It extracts signature genes that are variably expressed between cells and correlate to other genes.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">get_correlated_variable_genes =<span class="st"> </span><span class="cf">function</span>(mat, <span class="dt">n =</span> <span class="kw">nrow</span>(mat), <span class="dt">cor_cutoff =</span> <span class="dv">0</span>, <span class="dt">n_cutoff =</span> <span class="dv">0</span>) {
    ind =<span class="st"> </span><span class="kw">order</span>(<span class="kw">apply</span>(mat, <span class="dv">1</span>, <span class="cf">function</span>(x) {
            q =<span class="st"> </span><span class="kw">quantile</span>(x, <span class="kw">c</span>(<span class="fl">0.1</span>, <span class="fl">0.9</span>))
            x =<span class="st"> </span>x[x <span class="op">&lt;</span><span class="st"> </span>q[<span class="dv">1</span>] <span class="op">&amp;</span><span class="st"> </span>x <span class="op">&gt;</span><span class="st"> </span>q[<span class="dv">2</span>]]
            <span class="kw">var</span>(x)<span class="op">/</span><span class="kw">mean</span>(x)
        }), <span class="dt">decreasing =</span> <span class="ot">TRUE</span>)[<span class="dv">1</span><span class="op">:</span>n]
    mat2 =<span class="st"> </span>mat[ind, , drop =<span class="st"> </span><span class="ot">FALSE</span>]
    dt =<span class="st"> </span><span class="kw">cor</span>(<span class="kw">t</span>(mat2), <span class="dt">method =</span> <span class="st">&quot;spearman&quot;</span>)
    <span class="kw">diag</span>(dt) =<span class="st"> </span><span class="dv">0</span>
    dt[<span class="kw">abs</span>(dt) <span class="op">&lt;</span><span class="st"> </span>cor_cutoff] =<span class="st"> </span><span class="dv">0</span>
    dt[dt <span class="op">&lt;</span><span class="st"> </span><span class="dv">0</span>] =<span class="st"> </span><span class="op">-</span><span class="dv">1</span>
    dt[dt <span class="op">&gt;</span><span class="st"> </span><span class="dv">0</span>] =<span class="st"> </span><span class="dv">1</span>

    i =<span class="st"> </span><span class="kw">colSums</span>(<span class="kw">abs</span>(dt)) <span class="op">&gt;</span><span class="st"> </span>n_cutoff

    mat3 =<span class="st"> </span>mat2[i, ,drop =<span class="st"> </span><span class="ot">FALSE</span>]
    <span class="kw">return</span>(mat3)
}</code></pre></div>
<p>Signature genes are defined as a list of genes where each gene correlates to more than 20 genes with an absolute correlation larger than 0.5.</p>
<p><code>mat2</code> contains expression values scaled per gene, which means it contains relative expression across cells for every gene. Since single cell RNASeq data is highly variable and outliers are frequent, gene expression is only scaled within the 10th and 90th quantiles.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">mat =<span class="st"> </span><span class="kw">get_correlated_variable_genes</span>(expr, <span class="dt">cor_cutoff =</span> <span class="fl">0.5</span>, <span class="dt">n_cutoff =</span> <span class="dv">20</span>)
mat2 =<span class="st"> </span><span class="kw">t</span>(<span class="kw">apply</span>(mat, <span class="dv">1</span>, <span class="cf">function</span>(x) {
    q10 =<span class="st"> </span><span class="kw">quantile</span>(x, <span class="fl">0.1</span>)
    q90 =<span class="st"> </span><span class="kw">quantile</span>(x, <span class="fl">0.9</span>)
    x[x <span class="op">&lt;</span><span class="st"> </span>q10] =<span class="st"> </span>q10
    x[x <span class="op">&gt;</span><span class="st"> </span>q90] =<span class="st"> </span>q90
    <span class="kw">scale</span>(x)
}))
<span class="kw">colnames</span>(mat2) =<span class="st"> </span><span class="kw">colnames</span>(mat)</code></pre></div>
<p>Load cell cycle genes and ribonucleoprotein genes. The cell cycle gene list is from <a href="http://www.nature.com/nbt/journal/v33/n2/full/nbt.3102.html">Buettner et al., 2015</a>, supplementary table 1, sheet “<strong>Union of Cyclebase and GO genes</strong>”. Ribonucleoprotein genes are from <a href="http://amigo.geneontology.org/amigo/term/GO:0030529">GO:0030529</a>. Gene list are stored in <code>mouse_cell_cycle_gene.rds</code> and <code>mouse_ribonucleoprotein.rds</code>. The two files can be found <a href="https://jokergoo.github.io/ComplexHeatmap-reference/data/mouse_cell_cycle_gene.rds">here</a> and <a href="https://jokergoo.github.io/ComplexHeatmap-reference/data/mouse_ribonucleoprotein.rds">here</a>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">cc =<span class="st"> </span><span class="kw">readRDS</span>(<span class="st">&quot;data/mouse_cell_cycle_gene.rds&quot;</span>)
ccl =<span class="st"> </span><span class="kw">rownames</span>(mat) <span class="op">%in%</span><span class="st"> </span>cc
cc_gene =<span class="st"> </span><span class="kw">rownames</span>(mat)[ccl]

rp =<span class="st"> </span><span class="kw">readRDS</span>(<span class="st">&quot;data/mouse_ribonucleoprotein.rds&quot;</span>)
rpl =<span class="st"> </span><span class="kw">rownames</span>(mat) <span class="op">%in%</span><span class="st"> </span>rp</code></pre></div>
<p>Since with scaling the expression values per gene the expression level of a gene relative to other genes has been lost, we calculate the base mean as the mean expression of a gene throughout all samples. The base mean can be used to compare expression levels between genes.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">base_mean =<span class="st"> </span><span class="kw">rowMeans</span>(mat)</code></pre></div>
<p>Now the following information is available:</p>
<ol style="list-style-type: decimal">
<li>scaled expression, <code>mat2</code>,</li>
<li>base mean, <code>base_mean</code>,</li>
<li>whether genes are ribonucleoprotein genes, <code>rpl</code>,</li>
<li>whether genes are cell cycle genes, <code>ccl</code>,</li>
<li>symbols for cell cycle genes, <code>cc_gene</code>,</li>
</ol>
<p>In the next step, we can put the information together and visualize it as a list of heatmaps. A gene-gene correlation heatmap is added at the end and defined to be the main_heatmap, meaning that the row order of all heatmaps/row annotations are based on the clustering of this correlation matrix.</p>
<p>For cell cycle genes with relatively high expression levels (larger than the 25% quantile of all genes), the gene name is indicated as text labels. In the first heatmap, the column dendrogram is underlaid with two different colours based in the two main groups derived by hierarchical clustering to highlight the two subpopulations.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(GetoptLong)
ht_list =<span class="st"> </span><span class="kw">Heatmap</span>(mat2, <span class="dt">col =</span> <span class="kw">colorRamp2</span>(<span class="kw">c</span>(<span class="op">-</span><span class="fl">1.5</span>, <span class="dv">0</span>, <span class="fl">1.5</span>), <span class="kw">c</span>(<span class="st">&quot;blue&quot;</span>, <span class="st">&quot;white&quot;</span>, <span class="st">&quot;red&quot;</span>)), 
    <span class="dt">name =</span> <span class="st">&quot;scaled_expr&quot;</span>, <span class="dt">column_title =</span> <span class="kw">qq</span>(<span class="st">&quot;relative expression for @{nrow(mat)} genes&quot;</span>),
    <span class="dt">show_column_names =</span> <span class="ot">FALSE</span>, <span class="dt">width =</span> <span class="kw">unit</span>(<span class="dv">8</span>, <span class="st">&quot;cm&quot;</span>),
    <span class="dt">heatmap_legend_param =</span> <span class="kw">list</span>(<span class="dt">title =</span> <span class="st">&quot;Scaled expr&quot;</span>)) <span class="op">+</span>
<span class="st">    </span><span class="kw">Heatmap</span>(base_mean, <span class="dt">name =</span> <span class="st">&quot;base_expr&quot;</span>, <span class="dt">width =</span> <span class="kw">unit</span>(<span class="dv">5</span>, <span class="st">&quot;mm&quot;</span>),
        <span class="dt">heatmap_legend_param =</span> <span class="kw">list</span>(<span class="dt">title =</span> <span class="st">&quot;Base expr&quot;</span>)) <span class="op">+</span>
<span class="st">    </span><span class="kw">Heatmap</span>(rpl <span class="op">+</span><span class="st"> </span><span class="dv">0</span>, <span class="dt">name =</span> <span class="st">&quot;ribonucleoprotein&quot;</span>, <span class="dt">col =</span> <span class="kw">c</span>(<span class="st">&quot;0&quot;</span> =<span class="st"> &quot;white&quot;</span>, <span class="st">&quot;1&quot;</span> =<span class="st"> &quot;purple&quot;</span>), 
        <span class="dt">show_heatmap_legend =</span> <span class="ot">FALSE</span>, <span class="dt">width =</span> <span class="kw">unit</span>(<span class="dv">5</span>, <span class="st">&quot;mm&quot;</span>)) <span class="op">+</span>
<span class="st">    </span><span class="kw">Heatmap</span>(ccl <span class="op">+</span><span class="st"> </span><span class="dv">0</span>, <span class="dt">name =</span> <span class="st">&quot;cell_cycle&quot;</span>, <span class="dt">col =</span> <span class="kw">c</span>(<span class="st">&quot;0&quot;</span> =<span class="st"> &quot;white&quot;</span>, <span class="st">&quot;1&quot;</span> =<span class="st"> &quot;red&quot;</span>), 
        <span class="dt">show_heatmap_legend =</span> <span class="ot">FALSE</span>, <span class="dt">width =</span> <span class="kw">unit</span>(<span class="dv">5</span>, <span class="st">&quot;mm&quot;</span>)) <span class="op">+</span>
<span class="st">    </span><span class="kw">rowAnnotation</span>(<span class="dt">link =</span> <span class="kw">anno_mark</span>(<span class="dt">at =</span> <span class="kw">which</span>(ccl <span class="op">&amp;</span><span class="st"> </span>base_mean <span class="op">&gt;</span><span class="st"> </span><span class="kw">quantile</span>(base_mean, <span class="fl">0.25</span>)), 
        <span class="dt">labels =</span> <span class="kw">rownames</span>(mat)[ccl <span class="op">&amp;</span><span class="st"> </span>base_mean <span class="op">&gt;</span><span class="st"> </span><span class="kw">quantile</span>(base_mean, <span class="fl">0.25</span>)], 
        <span class="dt">labels_gp =</span> <span class="kw">gpar</span>(<span class="dt">fontsize =</span> <span class="dv">10</span>), <span class="dt">padding =</span> <span class="kw">unit</span>(<span class="dv">1</span>, <span class="st">&quot;mm&quot;</span>))) <span class="op">+</span>
<span class="st">    </span><span class="kw">Heatmap</span>(<span class="kw">cor</span>(<span class="kw">t</span>(mat2)), <span class="dt">name =</span> <span class="st">&quot;cor&quot;</span>, 
        <span class="dt">col =</span> <span class="kw">colorRamp2</span>(<span class="kw">c</span>(<span class="op">-</span><span class="dv">1</span>, <span class="dv">0</span>, <span class="dv">1</span>), <span class="kw">c</span>(<span class="st">&quot;green&quot;</span>, <span class="st">&quot;white&quot;</span>, <span class="st">&quot;red&quot;</span>)), 
        <span class="dt">show_row_names =</span> <span class="ot">FALSE</span>, <span class="dt">show_column_names =</span> <span class="ot">FALSE</span>, <span class="dt">row_dend_side =</span> <span class="st">&quot;right&quot;</span>, 
        <span class="dt">show_column_dend =</span> <span class="ot">FALSE</span>, <span class="dt">column_title =</span> <span class="st">&quot;pairwise correlation between genes&quot;</span>,
        <span class="dt">heatmap_legend_param =</span> <span class="kw">list</span>(<span class="dt">title =</span> <span class="st">&quot;Correlation&quot;</span>))
ht_list =<span class="st"> </span><span class="kw">draw</span>(ht_list, <span class="dt">main_heatmap =</span> <span class="st">&quot;cor&quot;</span>)
<span class="kw">decorate_column_dend</span>(<span class="st">&quot;scaled_expr&quot;</span>, {
    tree =<span class="st"> </span><span class="kw">column_dend</span>(ht_list)<span class="op">$</span>scaled_expr
    ind =<span class="st"> </span><span class="kw">cutree</span>(<span class="kw">as.hclust</span>(tree), <span class="dt">k =</span> <span class="dv">2</span>)[<span class="kw">order.dendrogram</span>(tree)]

    first_index =<span class="st"> </span><span class="cf">function</span>(l) <span class="kw">which</span>(l)[<span class="dv">1</span>]
    last_index =<span class="st"> </span><span class="cf">function</span>(l) { x =<span class="st"> </span><span class="kw">which</span>(l); x[<span class="kw">length</span>(x)] }
    x1 =<span class="st"> </span><span class="kw">c</span>(<span class="kw">first_index</span>(ind <span class="op">==</span><span class="st"> </span><span class="dv">1</span>), <span class="kw">first_index</span>(ind <span class="op">==</span><span class="st"> </span><span class="dv">2</span>)) <span class="op">-</span><span class="st"> </span><span class="dv">1</span>
    x2 =<span class="st"> </span><span class="kw">c</span>(<span class="kw">last_index</span>(ind <span class="op">==</span><span class="st"> </span><span class="dv">1</span>), <span class="kw">last_index</span>(ind <span class="op">==</span><span class="st"> </span><span class="dv">2</span>))
    <span class="kw">grid.rect</span>(<span class="dt">x =</span> x1<span class="op">/</span><span class="kw">length</span>(ind), <span class="dt">width =</span> (x2 <span class="op">-</span><span class="st"> </span>x1)<span class="op">/</span><span class="kw">length</span>(ind), <span class="dt">just =</span> <span class="st">&quot;left&quot;</span>,
        <span class="dt">default.units =</span> <span class="st">&quot;npc&quot;</span>, <span class="dt">gp =</span> <span class="kw">gpar</span>(<span class="dt">fill =</span> <span class="kw">c</span>(<span class="st">&quot;#FF000040&quot;</span>, <span class="st">&quot;#00FF0040&quot;</span>), <span class="dt">col =</span> <span class="ot">NA</span>))
})</code></pre></div>
<p><img src="09-examples_files/figure-html/unnamed-chunk-10-1.png" width="1152" style="display: block; margin: auto;" /></p>
<p>The heatmap clearly reveals that the cells are separated into two sub-populations. The population on the left in the first heatmap exhibits high expression of a subset of cell cycle genes (cell cycle genes are indicated in “<strong>cell_cycle</strong>” heatmap). However, the overall expression level for these genes is relatively low (see “<strong>base_expr</strong>” heatmap). The population on the right has higher expression in the other signature genes. Interestingly, the signature genes which are higher expressed in this subpopulation are enriched for genes coding for ribonucleoproteins (see “<strong>ribonucleoprotein</strong>” heatmap). A subset of the ribonucleoprotein genes shows strong coexpression (see correlation heatmap) and overall high expression levels (“<strong>base_expr</strong>” heatmap).</p>
</div>
<div id="correlations-between-methylation-expression-and-other-genomic-features" class="section level2">
<h2><span class="header-section-number">9.4</span> Correlations between methylation, expression and other genomic features</h2>
<p>In the following example, data is randomly generated based on patterns found in an unpublished analysis. First we load the data. <code>meth.rds</code> can be found <a href="https://jokergoo.github.io/ComplexHeatmap-reference/data/meth.rds">here</a>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">res_list =<span class="st"> </span><span class="kw">readRDS</span>(<span class="st">&quot;data/meth.rds&quot;</span>)
type =<span class="st"> </span>res_list<span class="op">$</span>type
mat_meth =<span class="st"> </span>res_list<span class="op">$</span>mat_meth
mat_expr =<span class="st"> </span>res_list<span class="op">$</span>mat_expr
direction =<span class="st"> </span>res_list<span class="op">$</span>direction
cor_pvalue =<span class="st"> </span>res_list<span class="op">$</span>cor_pvalue
gene_type =<span class="st"> </span>res_list<span class="op">$</span>gene_type
anno_gene =<span class="st"> </span>res_list<span class="op">$</span>anno_gene
dist =<span class="st"> </span>res_list<span class="op">$</span>dist
anno_enhancer =<span class="st"> </span>res_list<span class="op">$</span>anno_enhancer</code></pre></div>
<p>The different sources of information and corresponding variables are:</p>
<ol style="list-style-type: decimal">
<li><code>type</code>: the label which shows whether the sample is tumor or normal.</li>
<li><code>mat_meth</code>: a matrix in which rows correspond to differetially methylated regions (DMRs). The value in the matrix is the mean methylation level in the DMR in every sample.</li>
<li><code>mat_expr</code>: a matrix in which rows correspond to genes which are associated to the DMRs (i.e. the nearest gene to the DMR). The value in the matrix is the expression level for each gene in each sample. Expression is scaled for every gene across samples.</li>
<li><code>direction</code>: direction of the methylation change (hyper meaning higher methylation in tumor samples, hypo means lower methylation in tumor samples).</li>
<li><code>cor_pvalue</code>: p-value for the correlation test between methylation and expression of the associated gene.</li>
<li><code>gene_type</code>: type of the genes (e.g. protein coding genes or lincRNAs).</li>
<li><code>anno_gene</code>: annotation to the gene models (intergenic, intragenic or TSS).</li>
<li><code>dist</code>: distance from DMRs to TSS of the assiciated genes.</li>
<li><code>anno_enhancer</code>: fraction of the DMR that overlaps enhancers.</li>
</ol>
<p>The data only includes DMRs for which methylation and expression of the associated gene are negatively correlated.</p>
<p>The clustering of columns for the methylation matrix are calculated first so that columns in the expression matrix can be adjusted to have the same column order as in the methylation matrix.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">column_tree =<span class="st"> </span><span class="kw">hclust</span>(<span class="kw">dist</span>(<span class="kw">t</span>(mat_meth)))
column_order =<span class="st"> </span>column_tree<span class="op">$</span>order</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(RColorBrewer)
meth_col_fun =<span class="st"> </span><span class="kw">colorRamp2</span>(<span class="kw">c</span>(<span class="dv">0</span>, <span class="fl">0.5</span>, <span class="dv">1</span>), <span class="kw">c</span>(<span class="st">&quot;blue&quot;</span>, <span class="st">&quot;white&quot;</span>, <span class="st">&quot;red&quot;</span>))
direction_col =<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;hyper&quot;</span> =<span class="st"> &quot;red&quot;</span>, <span class="st">&quot;hypo&quot;</span> =<span class="st"> &quot;blue&quot;</span>)
expr_col_fun =<span class="st"> </span><span class="kw">colorRamp2</span>(<span class="kw">c</span>(<span class="op">-</span><span class="dv">2</span>, <span class="dv">0</span>, <span class="dv">2</span>), <span class="kw">c</span>(<span class="st">&quot;green&quot;</span>, <span class="st">&quot;white&quot;</span>, <span class="st">&quot;red&quot;</span>))
pvalue_col_fun =<span class="st"> </span><span class="kw">colorRamp2</span>(<span class="kw">c</span>(<span class="dv">0</span>, <span class="dv">2</span>, <span class="dv">4</span>), <span class="kw">c</span>(<span class="st">&quot;white&quot;</span>, <span class="st">&quot;white&quot;</span>, <span class="st">&quot;red&quot;</span>))
gene_type_col =<span class="st"> </span><span class="kw">structure</span>(<span class="kw">brewer.pal</span>(<span class="kw">length</span>(<span class="kw">unique</span>(gene_type)), <span class="st">&quot;Set3&quot;</span>), 
    <span class="dt">names =</span> <span class="kw">unique</span>(gene_type))
anno_gene_col =<span class="st"> </span><span class="kw">structure</span>(<span class="kw">brewer.pal</span>(<span class="kw">length</span>(<span class="kw">unique</span>(anno_gene)), <span class="st">&quot;Set1&quot;</span>), 
    <span class="dt">names =</span> <span class="kw">unique</span>(anno_gene))
dist_col_fun =<span class="st"> </span><span class="kw">colorRamp2</span>(<span class="kw">c</span>(<span class="dv">0</span>, <span class="dv">10000</span>), <span class="kw">c</span>(<span class="st">&quot;black&quot;</span>, <span class="st">&quot;white&quot;</span>))
enhancer_col_fun =<span class="st"> </span><span class="kw">colorRamp2</span>(<span class="kw">c</span>(<span class="dv">0</span>, <span class="dv">1</span>), <span class="kw">c</span>(<span class="st">&quot;white&quot;</span>, <span class="st">&quot;orange&quot;</span>))</code></pre></div>
<p>We first define two column annotations and then make the complex heatmaps.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">ht_opt</span>(
    <span class="dt">legend_title_gp =</span> <span class="kw">gpar</span>(<span class="dt">fontsize =</span> <span class="dv">8</span>, <span class="dt">fontface =</span> <span class="st">&quot;bold&quot;</span>), 
    <span class="dt">legend_labels_gp =</span> <span class="kw">gpar</span>(<span class="dt">fontsize =</span> <span class="dv">8</span>), 
    <span class="dt">heatmap_column_names_gp =</span> <span class="kw">gpar</span>(<span class="dt">fontsize =</span> <span class="dv">8</span>),
    <span class="dt">heatmap_column_title_gp =</span> <span class="kw">gpar</span>(<span class="dt">fontsize =</span> <span class="dv">10</span>),
    <span class="dt">heatmap_row_title_gp =</span> <span class="kw">gpar</span>(<span class="dt">fontsize =</span> <span class="dv">8</span>)
)

ha =<span class="st"> </span><span class="kw">HeatmapAnnotation</span>(<span class="dt">type =</span> type, 
    <span class="dt">col =</span> <span class="kw">list</span>(<span class="dt">type =</span> <span class="kw">c</span>(<span class="st">&quot;Tumor&quot;</span> =<span class="st"> &quot;pink&quot;</span>, <span class="st">&quot;Control&quot;</span> =<span class="st"> &quot;royalblue&quot;</span>)),
    <span class="dt">annotation_name_side =</span> <span class="st">&quot;left&quot;</span>)
ha2 =<span class="st"> </span><span class="kw">HeatmapAnnotation</span>(<span class="dt">type =</span> type, 
    <span class="dt">col =</span> <span class="kw">list</span>(<span class="dt">type =</span> <span class="kw">c</span>(<span class="st">&quot;Tumor&quot;</span> =<span class="st"> &quot;pink&quot;</span>, <span class="st">&quot;Control&quot;</span> =<span class="st"> &quot;royalblue&quot;</span>)), 
    <span class="dt">show_legend =</span> <span class="ot">FALSE</span>)

ht_list =<span class="st"> </span><span class="kw">Heatmap</span>(mat_meth, <span class="dt">name =</span> <span class="st">&quot;methylation&quot;</span>, <span class="dt">col =</span> meth_col_fun,
    <span class="dt">column_order=</span> column_order,
    <span class="dt">top_annotation =</span> ha, <span class="dt">column_title =</span> <span class="st">&quot;Methylation&quot;</span>) <span class="op">+</span>
<span class="st">    </span><span class="kw">Heatmap</span>(direction, <span class="dt">name =</span> <span class="st">&quot;direction&quot;</span>, <span class="dt">col =</span> direction_col) <span class="op">+</span>
<span class="st">    </span><span class="kw">Heatmap</span>(mat_expr[, column_tree<span class="op">$</span>order], <span class="dt">name =</span> <span class="st">&quot;expression&quot;</span>, 
        <span class="dt">col =</span> expr_col_fun, 
        <span class="dt">column_order =</span> column_order, 
        <span class="dt">top_annotation =</span> ha2, <span class="dt">column_title =</span> <span class="st">&quot;Expression&quot;</span>) <span class="op">+</span>
<span class="st">    </span><span class="kw">Heatmap</span>(cor_pvalue, <span class="dt">name =</span> <span class="st">&quot;-log10(cor_p)&quot;</span>, <span class="dt">col =</span> pvalue_col_fun) <span class="op">+</span>
<span class="st">    </span><span class="kw">Heatmap</span>(gene_type, <span class="dt">name =</span> <span class="st">&quot;gene type&quot;</span>, <span class="dt">col =</span> gene_type_col) <span class="op">+</span>
<span class="st">    </span><span class="kw">Heatmap</span>(anno_gene, <span class="dt">name =</span> <span class="st">&quot;anno_gene&quot;</span>, <span class="dt">col =</span> anno_gene_col) <span class="op">+</span>
<span class="st">    </span><span class="kw">Heatmap</span>(dist, <span class="dt">name =</span> <span class="st">&quot;dist_tss&quot;</span>, <span class="dt">col =</span> dist_col_fun) <span class="op">+</span>
<span class="st">    </span><span class="kw">Heatmap</span>(anno_enhancer, <span class="dt">name =</span> <span class="st">&quot;anno_enhancer&quot;</span>, <span class="dt">col =</span> enhancer_col_fun, 
        <span class="dt">cluster_columns =</span> <span class="ot">FALSE</span>, <span class="dt">column_title =</span> <span class="st">&quot;Enhancer&quot;</span>)

<span class="kw">draw</span>(ht_list, <span class="dt">row_km =</span> <span class="dv">2</span>, <span class="dt">row_split =</span> direction,
    <span class="dt">column_title =</span> <span class="st">&quot;Comprehensive correspondence between methylation, expression and other genomic features&quot;</span>, 
    <span class="dt">column_title_gp =</span> <span class="kw">gpar</span>(<span class="dt">fontsize =</span> <span class="dv">12</span>, <span class="dt">fontface =</span> <span class="st">&quot;bold&quot;</span>), 
    <span class="dt">merge_legends =</span> <span class="ot">TRUE</span>, <span class="dt">heatmap_legend_side =</span> <span class="st">&quot;bottom&quot;</span>)</code></pre></div>
<p><img src="09-examples_files/figure-html/unnamed-chunk-14-1.png" width="1152" style="display: block; margin: auto;" /></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">ht_opt</span>(<span class="dt">RESET =</span> <span class="ot">TRUE</span>)</code></pre></div>
<p>The complex heatmaps reveal that highly methylated DMRs are enriched in intergenic and intragenic regions and rarely overlap with enhancers. In contrast, lowly methylated DMRs are enriched for transcription start sites (TSS) and enhancers.</p>
</div>
<div id="visualize-methylation-profile-with-complex-annotations" class="section level2">
<h2><span class="header-section-number">9.5</span> Visualize Methylation Profile with Complex Annotations</h2>
<p>In this example, Figure 1 in <a href="http://dx.doi.org/10.1016/j.ccr.2012.08.024">Strum et al., 2012</a> is re-implemented with some adjustments.</p>
<p>Some packages need to be loaded firstly.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(matrixStats)
<span class="kw">library</span>(GenomicRanges)</code></pre></div>
<p>Methylation profiles can be download from <a href="http://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE36278">GEO database</a>. The <a href="http://bioconductor.org/packages/release/bioc/html/GEOquery.html"><strong>GEOquery</strong> package</a> is used to retrieve data from GEO.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(GEOquery)
gset =<span class="st"> </span><span class="kw">getGEO</span>(<span class="st">&quot;GSE36278&quot;</span>)</code></pre></div>
<p>The methylation profiles have been measured by Illumina HumanMethylation450 BeadChip arrays. We load probe data via the <strong>IlluminaHumanMethylation450kanno.ilmn12.hg19</strong> package.</p>
<p>Adjust row names in the matrix to be the same as the probes.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(IlluminaHumanMethylation450kanno.ilmn12.hg19)
<span class="kw">data</span>(Locations)

mat =<span class="st"> </span><span class="kw">exprs</span>(gset[[<span class="dv">1</span>]])
<span class="kw">colnames</span>(mat) =<span class="st"> </span><span class="kw">phenoData</span>(gset[[<span class="dv">1</span>]])<span class="op">@</span>data<span class="op">$</span>title
mat =<span class="st"> </span>mat[<span class="kw">rownames</span>(Locations), ] </code></pre></div>
<p><code>probe</code> contains locations of probes and also information whether the CpG sites overlap with SNPs. Here we remove probes that are on sex chromosomes and probes that overlap with SNPs.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">data</span>(SNPs.137CommonSingle)
<span class="kw">data</span>(Islands.UCSC)
l =<span class="st"> </span>Locations<span class="op">$</span>chr <span class="op">%in%</span><span class="st"> </span><span class="kw">paste0</span>(<span class="st">&quot;chr&quot;</span>, <span class="dv">1</span><span class="op">:</span><span class="dv">22</span>) <span class="op">&amp;</span><span class="st"> </span><span class="kw">is.na</span>(SNPs.137CommonSingle<span class="op">$</span>Probe_rs)
mat =<span class="st"> </span>mat[l, ]</code></pre></div>
<p>Get subsets for locations of probes and the annotation to CpG Islands accordingly.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">cgi =<span class="st"> </span>Islands.UCSC<span class="op">$</span>Relation_to_Island[l]
loc =<span class="st"> </span>Locations[l, ]</code></pre></div>
<p>Separate the matrix into a matrix for tumor samples and a matrix for normal samples. Also modify column names for the tumor samples to be consistent with the phenotype data which we will read later.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">mat1 =<span class="st"> </span><span class="kw">as.matrix</span>(mat[, <span class="kw">grep</span>(<span class="st">&quot;GBM&quot;</span>, <span class="kw">colnames</span>(mat))])   <span class="co"># tumor samples</span>
mat2 =<span class="st"> </span><span class="kw">as.matrix</span>(mat[, <span class="kw">grep</span>(<span class="st">&quot;CTRL&quot;</span>, <span class="kw">colnames</span>(mat))])  <span class="co"># normal samples</span>
<span class="kw">colnames</span>(mat1) =<span class="st"> </span><span class="kw">gsub</span>(<span class="st">&quot;GBM&quot;</span>, <span class="st">&quot;dkfz&quot;</span>, <span class="kw">colnames</span>(mat1))</code></pre></div>
<p>Phenotype data is from <a href="http://dx.doi.org/10.1016/j.ccr.2012.08.024">Sturm et al., 2012</a>, supplementary table S1 and can be found <a href="https://jokergoo.github.io/ComplexHeatmap-reference/data/450K_annotation.txt">here</a>.</p>
<p>The rows of phenotype data are adjusted to be the same as the columns of the methylation matrix.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">phenotype =<span class="st"> </span><span class="kw">read.table</span>(<span class="st">&quot;data/450K_annotation.txt&quot;</span>, <span class="dt">header =</span> <span class="ot">TRUE</span>, <span class="dt">sep =</span> <span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>, 
    <span class="dt">row.names =</span> <span class="dv">1</span>, <span class="dt">check.names =</span> <span class="ot">FALSE</span>, <span class="dt">comment.char =</span> <span class="st">&quot;&quot;</span>, <span class="dt">stringsAsFactors =</span> <span class="ot">FALSE</span>)
phenotype =<span class="st"> </span>phenotype[<span class="kw">colnames</span>(mat1), ]</code></pre></div>
<p>Please note that we only use the 136 samples which are from DKFZ, while in <a href="http://dx.doi.org/10.1016/j.ccr.2012.08.024">Sturm et al., 2012</a>, additional 74 TCGA samples have been used.</p>
<p>Extract the top 8000 probes with most variable methylation in the tumor samples, and also subset other information correspondingly.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">ind =<span class="st"> </span><span class="kw">order</span>(<span class="kw">rowVars</span>(mat1, <span class="dt">na.rm =</span> <span class="ot">TRUE</span>), <span class="dt">decreasing =</span> <span class="ot">TRUE</span>)[<span class="dv">1</span><span class="op">:</span><span class="dv">8000</span>]
m1 =<span class="st"> </span>mat1[ind, ]
m2 =<span class="st"> </span>mat2[ind, ]
cgi2 =<span class="st"> </span>cgi[ind]
cgi2 =<span class="st"> </span><span class="kw">ifelse</span>(<span class="kw">grepl</span>(<span class="st">&quot;Shore&quot;</span>, cgi2), <span class="st">&quot;Shore&quot;</span>, cgi2)
cgi2 =<span class="st"> </span><span class="kw">ifelse</span>(<span class="kw">grepl</span>(<span class="st">&quot;Shelf&quot;</span>, cgi2), <span class="st">&quot;Shelf&quot;</span>, cgi2)
loc =<span class="st"> </span>loc[ind, ]</code></pre></div>
<p>For each probe, find the distance to the closest TSS. <code>pc_tx_tss.bed</code> contains positions of TSS from protein coding genes.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">gr =<span class="st"> </span><span class="kw">GRanges</span>(loc[, <span class="dv">1</span>], <span class="dt">ranges =</span> <span class="kw">IRanges</span>(loc[, <span class="dv">2</span>], loc[, <span class="dv">2</span>]<span class="op">+</span><span class="dv">1</span>))
tss =<span class="st"> </span><span class="kw">read.table</span>(<span class="st">&quot;data/pc_tx_tss.bed&quot;</span>, <span class="dt">stringsAsFactors =</span> <span class="ot">FALSE</span>)
tss =<span class="st"> </span><span class="kw">GRanges</span>(tss[[<span class="dv">1</span>]], <span class="dt">ranges =</span> <span class="kw">IRanges</span>(tss[, <span class="dv">2</span>], tss[, <span class="dv">3</span>]))

tss_dist =<span class="st"> </span><span class="kw">distanceToNearest</span>(gr, tss)
tss_dist =<span class="st"> </span>tss_dist<span class="op">@</span>elementMetadata<span class="op">$</span>distance</code></pre></div>
<p>Because there are a few <code>NA</code> in the matrix (<code>sum(is.na(m1))/length(m1)</code> = 0.0011967) which will break the <code>cor()</code> function, we replace <code>NA</code> to the intermediate methylation (0.5). Note that although <strong>ComplexHeatmap</strong> allows <code>NA</code> in the matrix, removal of <code>NA</code> will speed up the clustering.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">m1[<span class="kw">is.na</span>(m1)] =<span class="st"> </span><span class="fl">0.5</span>
m2[<span class="kw">is.na</span>(m2)] =<span class="st"> </span><span class="fl">0.5</span></code></pre></div>
<p>The following annotations will be added to the columns of the methylation matrix:</p>
<ol style="list-style-type: decimal">
<li>age</li>
<li>subtype classification by DKFZ</li>
<li>subtype classification by TCGA</li>
<li>subtype classification by TCGA, based on expression profile</li>
<li>IDH1 mutation</li>
<li>H3F3A mutation</li>
<li>TP53 mutation</li>
<li>chr7 gain</li>
<li>chr10 loss</li>
<li>CDKN2A deletion</li>
<li>EGFR amplification</li>
<li>PDGFRA amplification</li>
</ol>
<p>In following code we define the column annotation in the <code>ha</code> variable. Also we customize colors, legends and height of the annotations.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">mutation_col =<span class="st"> </span><span class="kw">structure</span>(<span class="dt">names =</span> <span class="kw">c</span>(<span class="st">&quot;MUT&quot;</span>, <span class="st">&quot;WT&quot;</span>, <span class="st">&quot;G34R&quot;</span>, <span class="st">&quot;G34V&quot;</span>, <span class="st">&quot;K27M&quot;</span>), 
    <span class="kw">c</span>(<span class="st">&quot;black&quot;</span>, <span class="st">&quot;white&quot;</span>, <span class="st">&quot;#4DAF4A&quot;</span>, <span class="st">&quot;#4DAF4A&quot;</span>, <span class="st">&quot;#377EB8&quot;</span>))
cnv_col =<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;gain&quot;</span> =<span class="st"> &quot;#E41A1C&quot;</span>, <span class="st">&quot;loss&quot;</span> =<span class="st"> &quot;#377EB8&quot;</span>, <span class="st">&quot;amp&quot;</span> =<span class="st"> &quot;#E41A1C&quot;</span>, 
    <span class="st">&quot;del&quot;</span> =<span class="st"> &quot;#377EB8&quot;</span>, <span class="st">&quot;normal&quot;</span> =<span class="st"> &quot;white&quot;</span>)
ha =<span class="st"> </span><span class="kw">HeatmapAnnotation</span>(
    <span class="dt">age =</span> <span class="kw">anno_points</span>(phenotype[[<span class="dv">13</span>]], 
        <span class="dt">gp =</span> <span class="kw">gpar</span>(<span class="dt">col =</span> <span class="kw">ifelse</span>(phenotype[[<span class="dv">13</span>]] <span class="op">&gt;</span><span class="st"> </span><span class="dv">20</span>, <span class="st">&quot;black&quot;</span>, <span class="st">&quot;red&quot;</span>)), 
        <span class="dt">height =</span> <span class="kw">unit</span>(<span class="dv">3</span>, <span class="st">&quot;cm&quot;</span>)),
    <span class="dt">dkfz_cluster =</span> phenotype[[<span class="dv">1</span>]],
    <span class="dt">tcga_cluster =</span> phenotype[[<span class="dv">2</span>]],
    <span class="dt">tcga_expr =</span> phenotype[[<span class="dv">3</span>]],
    <span class="dt">IDH1 =</span> phenotype[[<span class="dv">5</span>]],
    <span class="dt">H3F3A =</span> phenotype[[<span class="dv">4</span>]],
    <span class="dt">TP53 =</span> phenotype[[<span class="dv">6</span>]],
    <span class="dt">chr7_gain =</span> <span class="kw">ifelse</span>(phenotype[[<span class="dv">7</span>]] <span class="op">==</span><span class="st"> </span><span class="dv">1</span>, <span class="st">&quot;gain&quot;</span>, <span class="st">&quot;normal&quot;</span>),
    <span class="dt">chr10_loss =</span> <span class="kw">ifelse</span>(phenotype[[<span class="dv">8</span>]] <span class="op">==</span><span class="st"> </span><span class="dv">1</span>, <span class="st">&quot;loss&quot;</span>, <span class="st">&quot;normal&quot;</span>),
    <span class="dt">CDKN2A_del =</span> <span class="kw">ifelse</span>(phenotype[[<span class="dv">9</span>]] <span class="op">==</span><span class="st"> </span><span class="dv">1</span>, <span class="st">&quot;del&quot;</span>, <span class="st">&quot;normal&quot;</span>),
    <span class="dt">EGFR_amp =</span> <span class="kw">ifelse</span>(phenotype[[<span class="dv">10</span>]] <span class="op">==</span><span class="st"> </span><span class="dv">1</span>, <span class="st">&quot;amp&quot;</span>, <span class="st">&quot;normal&quot;</span>),
    <span class="dt">PDGFRA_amp =</span> <span class="kw">ifelse</span>(phenotype[[<span class="dv">11</span>]] <span class="op">==</span><span class="st"> </span><span class="dv">1</span>, <span class="st">&quot;amp&quot;</span>, <span class="st">&quot;normal&quot;</span>),
    <span class="dt">col =</span> <span class="kw">list</span>(<span class="dt">dkfz_cluster =</span> <span class="kw">structure</span>(<span class="dt">names =</span> <span class="kw">c</span>(<span class="st">&quot;IDH&quot;</span>, <span class="st">&quot;K27&quot;</span>, <span class="st">&quot;G34&quot;</span>, <span class="st">&quot;RTK I PDGFRA&quot;</span>, 
            <span class="st">&quot;Mesenchymal&quot;</span>, <span class="st">&quot;RTK II Classic&quot;</span>), <span class="kw">brewer.pal</span>(<span class="dv">6</span>, <span class="st">&quot;Set1&quot;</span>)),
        <span class="dt">tcga_cluster =</span> <span class="kw">structure</span>(<span class="dt">names =</span> <span class="kw">c</span>(<span class="st">&quot;G-CIMP+&quot;</span>, <span class="st">&quot;Cluster #2&quot;</span>, <span class="st">&quot;Cluster #3&quot;</span>), 
            <span class="kw">brewer.pal</span>(<span class="dv">3</span>, <span class="st">&quot;Set1&quot;</span>)),
        <span class="dt">tcga_expr =</span> <span class="kw">structure</span>(<span class="dt">names =</span> <span class="kw">c</span>(<span class="st">&quot;Proneural&quot;</span>, <span class="st">&quot;Classical&quot;</span>, <span class="st">&quot;Mesenchymal&quot;</span>), 
            <span class="kw">c</span>(<span class="st">&quot;#377EB8&quot;</span>, <span class="st">&quot;#FFFF33&quot;</span>, <span class="st">&quot;#FF7F00&quot;</span>)),
        <span class="dt">IDH1 =</span> mutation_col,
        <span class="dt">H3F3A =</span> mutation_col,
        <span class="dt">TP53 =</span> mutation_col,
        <span class="dt">chr7_gain =</span> cnv_col,
        <span class="dt">chr10_loss =</span> cnv_col,
        <span class="dt">CDKN2A_del =</span> cnv_col,
        <span class="dt">EGFR_amp =</span> cnv_col,
        <span class="dt">PDGFRA_amp =</span> cnv_col),
    <span class="dt">na_col =</span> <span class="st">&quot;grey&quot;</span>, <span class="dt">border =</span> <span class="ot">TRUE</span>,
    <span class="dt">show_legend =</span> <span class="kw">c</span>(<span class="ot">TRUE</span>, <span class="ot">TRUE</span>, <span class="ot">TRUE</span>, <span class="ot">FALSE</span>, <span class="ot">TRUE</span>, <span class="ot">FALSE</span>, <span class="ot">TRUE</span>, <span class="ot">FALSE</span>, <span class="ot">FALSE</span>, <span class="ot">FALSE</span>, <span class="ot">FALSE</span>),
    <span class="dt">show_annotation_name =</span> <span class="ot">FALSE</span>,
    <span class="dt">annotation_legend_param =</span> <span class="kw">list</span>(
        <span class="dt">dkfz_cluster =</span> <span class="kw">list</span>(<span class="dt">title =</span> <span class="st">&quot;DKFZ Methylation&quot;</span>),
        <span class="dt">tcga_cluster =</span> <span class="kw">list</span>(<span class="dt">title =</span> <span class="st">&quot;TCGA Methylation&quot;</span>),
        <span class="dt">tcga_expr =</span> <span class="kw">list</span>(<span class="dt">title =</span> <span class="st">&quot;TCGA Expression&quot;</span>),
        <span class="dt">H3F3A =</span> <span class="kw">list</span>(<span class="dt">title =</span> <span class="st">&quot;Mutations&quot;</span>),
        <span class="dt">chr7_gain =</span> <span class="kw">list</span>(<span class="dt">title =</span> <span class="st">&quot;CNV&quot;</span>))
)</code></pre></div>
<p>In the final plot, there are four heatmaps added. From left to right, there are</p>
<ol style="list-style-type: decimal">
<li>heatmap for methylation in tumor samples</li>
<li>methylation in normal samples</li>
<li>distance to nearest TSS</li>
<li>CpG Island (CGI) annotation.</li>
</ol>
<p>The heatmaps are split by rows according to CGI annotations.</p>
<p>After the heatmaps are plotted, additional graphics such as labels for annotations are added by <code>decorate_*()</code> functions.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">col_fun =<span class="st"> </span><span class="kw">colorRamp2</span>(<span class="kw">c</span>(<span class="dv">0</span>, <span class="fl">0.5</span>, <span class="dv">1</span>), <span class="kw">c</span>(<span class="st">&quot;#377EB8&quot;</span>, <span class="st">&quot;white&quot;</span>, <span class="st">&quot;#E41A1C&quot;</span>))
ht_list =<span class="st"> </span><span class="kw">Heatmap</span>(m1, <span class="dt">col =</span> col_fun, <span class="dt">name =</span> <span class="st">&quot;Methylation&quot;</span>,
    <span class="dt">clustering_distance_columns =</span> <span class="st">&quot;spearman&quot;</span>,
    <span class="dt">show_row_dend =</span> <span class="ot">FALSE</span>, <span class="dt">show_column_dend =</span> <span class="ot">FALSE</span>,
    <span class="dt">show_column_names =</span> <span class="ot">FALSE</span>,
    <span class="dt">bottom_annotation =</span> ha, <span class="dt">column_title =</span> <span class="kw">qq</span>(<span class="st">&quot;GBM samples (n = @{ncol(m1)})&quot;</span>),
    <span class="dt">row_split =</span> <span class="kw">factor</span>(cgi2, <span class="dt">levels =</span> <span class="kw">c</span>(<span class="st">&quot;Island&quot;</span>, <span class="st">&quot;Shore&quot;</span>, <span class="st">&quot;Shelf&quot;</span>, <span class="st">&quot;OpenSea&quot;</span>)), 
    <span class="dt">row_title_gp =</span> <span class="kw">gpar</span>(<span class="dt">col =</span> <span class="st">&quot;#FFFFFF00&quot;</span>)) <span class="op">+</span><span class="st"> </span>
<span class="kw">Heatmap</span>(m2, <span class="dt">col =</span> col_fun, <span class="dt">show_column_names =</span> <span class="ot">FALSE</span>, 
    <span class="dt">show_column_dend =</span> <span class="ot">FALSE</span>, <span class="dt">column_title =</span> <span class="st">&quot;Controls&quot;</span>,
    <span class="dt">show_heatmap_legend =</span> <span class="ot">FALSE</span>, <span class="dt">width =</span> <span class="kw">unit</span>(<span class="dv">1</span>, <span class="st">&quot;cm&quot;</span>)) <span class="op">+</span>
<span class="kw">Heatmap</span>(tss_dist, <span class="dt">name =</span> <span class="st">&quot;tss_dist&quot;</span>, <span class="dt">col =</span> <span class="kw">colorRamp2</span>(<span class="kw">c</span>(<span class="dv">0</span>, <span class="fl">2e5</span>), <span class="kw">c</span>(<span class="st">&quot;white&quot;</span>, <span class="st">&quot;black&quot;</span>)), 
    <span class="dt">width =</span> <span class="kw">unit</span>(<span class="dv">5</span>, <span class="st">&quot;mm&quot;</span>),
    <span class="dt">heatmap_legend_param =</span> <span class="kw">list</span>(<span class="dt">at =</span> <span class="kw">c</span>(<span class="dv">0</span>, <span class="fl">1e5</span>, <span class="fl">2e5</span>), <span class="dt">labels =</span> <span class="kw">c</span>(<span class="st">&quot;0kb&quot;</span>, <span class="st">&quot;100kb&quot;</span>, <span class="st">&quot;200kb&quot;</span>))) <span class="op">+</span><span class="st"> </span>
<span class="kw">Heatmap</span>(cgi2, <span class="dt">name =</span> <span class="st">&quot;CGI&quot;</span>, <span class="dt">show_row_names =</span> <span class="ot">FALSE</span>, <span class="dt">width =</span> <span class="kw">unit</span>(<span class="dv">5</span>, <span class="st">&quot;mm&quot;</span>),
    <span class="dt">col =</span> <span class="kw">structure</span>(<span class="dt">names =</span> <span class="kw">c</span>(<span class="st">&quot;Island&quot;</span>, <span class="st">&quot;Shore&quot;</span>, <span class="st">&quot;Shelf&quot;</span>, <span class="st">&quot;OpenSea&quot;</span>), <span class="kw">c</span>(<span class="st">&quot;red&quot;</span>, <span class="st">&quot;blue&quot;</span>, <span class="st">&quot;green&quot;</span>, <span class="st">&quot;#CCCCCC&quot;</span>)))
<span class="kw">draw</span>(ht_list, <span class="dt">row_title =</span> <span class="kw">paste0</span>(<span class="st">&quot;DNA methylation probes (n = &quot;</span>, <span class="kw">nrow</span>(m1), <span class="st">&quot;)&quot;</span>),
    <span class="dt">annotation_legend_side =</span> <span class="st">&quot;left&quot;</span>, <span class="dt">heatmap_legend_side =</span> <span class="st">&quot;left&quot;</span>)

annotation_titles =<span class="st"> </span><span class="kw">c</span>(<span class="dt">dkfz_cluster =</span> <span class="st">&quot;DKFZ Methylation&quot;</span>,
    <span class="dt">tcga_cluster =</span> <span class="st">&quot;TCGA Methylation&quot;</span>,
    <span class="dt">tcga_expr =</span> <span class="st">&quot;TCGA Expression&quot;</span>,
    <span class="dt">IDH1 =</span> <span class="st">&quot;IDH1&quot;</span>,
    <span class="dt">H3F3A =</span> <span class="st">&quot;H3F3A&quot;</span>,
    <span class="dt">TP53 =</span> <span class="st">&quot;TP53&quot;</span>,
    <span class="dt">chr7_gain =</span> <span class="st">&quot;Chr7 gain&quot;</span>,
    <span class="dt">chr10_loss =</span> <span class="st">&quot;Chr10 loss&quot;</span>,
    <span class="dt">CDKN2A_del =</span> <span class="st">&quot;Chr10 loss&quot;</span>,
    <span class="dt">EGFR_amp =</span> <span class="st">&quot;EGFR amp&quot;</span>,
    <span class="dt">PDGFRA_amp =</span> <span class="st">&quot;PDGFRA amp&quot;</span>)
<span class="cf">for</span>(an <span class="cf">in</span> <span class="kw">names</span>(annotation_titles)) {
    <span class="kw">decorate_annotation</span>(an, {
        <span class="kw">grid.text</span>(annotation_titles[an], <span class="kw">unit</span>(<span class="op">-</span><span class="dv">2</span>, <span class="st">&quot;mm&quot;</span>), <span class="dt">just =</span> <span class="st">&quot;right&quot;</span>)
        <span class="kw">grid.rect</span>(<span class="dt">gp =</span> <span class="kw">gpar</span>(<span class="dt">fill =</span> <span class="ot">NA</span>, <span class="dt">col =</span> <span class="st">&quot;black&quot;</span>))
    })
}
<span class="kw">decorate_annotation</span>(<span class="st">&quot;age&quot;</span>, {
    <span class="kw">grid.text</span>(<span class="st">&quot;Age&quot;</span>, <span class="kw">unit</span>(<span class="dv">8</span>, <span class="st">&quot;mm&quot;</span>), <span class="dt">just =</span> <span class="st">&quot;right&quot;</span>)
    <span class="kw">grid.rect</span>(<span class="dt">gp =</span> <span class="kw">gpar</span>(<span class="dt">fill =</span> <span class="ot">NA</span>, <span class="dt">col =</span> <span class="st">&quot;black&quot;</span>))
    <span class="kw">grid.lines</span>(<span class="kw">unit</span>(<span class="kw">c</span>(<span class="dv">0</span>, <span class="dv">1</span>), <span class="st">&quot;npc&quot;</span>), <span class="kw">unit</span>(<span class="kw">c</span>(<span class="dv">20</span>, <span class="dv">20</span>), <span class="st">&quot;native&quot;</span>), <span class="dt">gp =</span> <span class="kw">gpar</span>(<span class="dt">lty =</span> <span class="dv">2</span>))
})
<span class="kw">decorate_annotation</span>(<span class="st">&quot;IDH1&quot;</span>, {
    <span class="kw">grid.lines</span>(<span class="kw">unit</span>(<span class="kw">c</span>(<span class="op">-</span><span class="dv">40</span>, <span class="dv">0</span>), <span class="st">&quot;mm&quot;</span>), <span class="kw">unit</span>(<span class="kw">c</span>(<span class="dv">1</span>, <span class="dv">1</span>), <span class="st">&quot;npc&quot;</span>))
})
<span class="kw">decorate_annotation</span>(<span class="st">&quot;chr7_gain&quot;</span>, {
    <span class="kw">grid.lines</span>(<span class="kw">unit</span>(<span class="kw">c</span>(<span class="op">-</span><span class="dv">40</span>, <span class="dv">0</span>), <span class="st">&quot;mm&quot;</span>), <span class="kw">unit</span>(<span class="kw">c</span>(<span class="dv">1</span>, <span class="dv">1</span>), <span class="st">&quot;npc&quot;</span>))
})</code></pre></div>
<p><img src="09-examples_files/figure-html/unnamed-chunk-27-1.png" width="1152" style="display: block; margin: auto;" /></p>
</div>
<div id="add-multiple-boxplots-for-single-row" class="section level2">
<h2><span class="header-section-number">9.6</span> Add multiple boxplots for single row</h2>
<p>The annotation function <code>anno_boxplot()</code> only draws one boxplot for a single row. When multiple heatmaps are concatenated, or the groups for columns have already been defined, for each row, we want to compare between heatmaps or column groups, thus, multiple boxplots need to be drawn for each single row.</p>
<p>In following example, we demonstrate how to implement an annotation function which draws multiple boxplots for single rows. The <code>grid.boxplot()</code> function is from <strong>ComplexHeatmap</strong> package which makes it easy to draw boxplot under <strong>grid</strong> system.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">m1 =<span class="st"> </span><span class="kw">matrix</span>(<span class="kw">sort</span>(<span class="kw">rnorm</span>(<span class="dv">100</span>)), <span class="dv">10</span>, <span class="dt">byrow =</span> <span class="ot">TRUE</span>)
m2 =<span class="st"> </span><span class="kw">matrix</span>(<span class="kw">sort</span>(<span class="kw">rnorm</span>(<span class="dv">100</span>), <span class="dt">decreasing =</span> <span class="ot">TRUE</span>), <span class="dv">10</span>, <span class="dt">byrow =</span> <span class="ot">TRUE</span>)
nr =<span class="st"> </span><span class="kw">nrow</span>(m1)

ht_list =<span class="st"> </span><span class="kw">Heatmap</span>(m1, <span class="dt">name =</span> <span class="st">&quot;m1&quot;</span>) <span class="op">+</span><span class="st"> </span><span class="kw">Heatmap</span>(m2, <span class="dt">name =</span> <span class="st">&quot;m2&quot;</span>)

rg =<span class="st"> </span><span class="kw">range</span>(<span class="kw">c</span>(m1, m2))
rg[<span class="dv">1</span>] =<span class="st"> </span>rg[<span class="dv">1</span>] <span class="op">-</span><span class="st"> </span>(rg[<span class="dv">2</span>] <span class="op">-</span><span class="st"> </span>rg[<span class="dv">1</span>])<span class="op">*</span><span class="st"> </span><span class="fl">0.02</span>
rg[<span class="dv">2</span>] =<span class="st"> </span>rg[<span class="dv">2</span>] <span class="op">+</span><span class="st"> </span>(rg[<span class="dv">2</span>] <span class="op">-</span><span class="st"> </span>rg[<span class="dv">1</span>])<span class="op">*</span><span class="st"> </span><span class="fl">0.02</span>
anno_multiple_boxplot =<span class="st"> </span><span class="cf">function</span>(index) {
    <span class="kw">pushViewport</span>(<span class="kw">viewport</span>(<span class="dt">xscale =</span> rg, <span class="dt">yscale =</span> <span class="kw">c</span>(<span class="fl">0.5</span>, nr <span class="op">+</span><span class="st"> </span><span class="fl">0.5</span>)))
    <span class="cf">for</span>(i <span class="cf">in</span> <span class="kw">seq_along</span>(index)) {
        <span class="kw">grid.rect</span>(<span class="dt">y =</span> nr<span class="op">-</span>i<span class="op">+</span><span class="dv">1</span>, <span class="dt">height =</span> <span class="dv">1</span>, <span class="dt">default.units =</span> <span class="st">&quot;native&quot;</span>)
        <span class="kw">grid.boxplot</span>(m1[ index[i], ], <span class="dt">pos =</span> nr<span class="op">-</span>i<span class="op">+</span><span class="dv">1</span> <span class="op">+</span><span class="st"> </span><span class="fl">0.2</span>, <span class="dt">box_width =</span> <span class="fl">0.3</span>, 
            <span class="dt">gp =</span> <span class="kw">gpar</span>(<span class="dt">fill =</span> <span class="st">&quot;red&quot;</span>), <span class="dt">direction =</span> <span class="st">&quot;horizontal&quot;</span>)
        <span class="kw">grid.boxplot</span>(m2[ index[i], ], <span class="dt">pos =</span> nr<span class="op">-</span>i<span class="op">+</span><span class="dv">1</span> <span class="op">-</span><span class="st"> </span><span class="fl">0.2</span>, <span class="dt">box_width =</span> <span class="fl">0.3</span>, 
            <span class="dt">gp =</span> <span class="kw">gpar</span>(<span class="dt">fill =</span> <span class="st">&quot;green&quot;</span>), <span class="dt">direction =</span> <span class="st">&quot;horizontal&quot;</span>)
    }
    <span class="kw">grid.xaxis</span>()
    <span class="kw">popViewport</span>()
}

ht_list =<span class="st"> </span>ht_list <span class="op">+</span><span class="st"> </span><span class="kw">rowAnnotation</span>(<span class="dt">boxplot =</span> anno_multiple_boxplot, <span class="dt">width =</span> <span class="kw">unit</span>(<span class="dv">4</span>, <span class="st">&quot;cm&quot;</span>), 
    <span class="dt">show_annotation_name =</span> <span class="ot">FALSE</span>)
lgd =<span class="st"> </span><span class="kw">Legend</span>(<span class="dt">labels =</span> <span class="kw">c</span>(<span class="st">&quot;m1&quot;</span>, <span class="st">&quot;m2&quot;</span>), <span class="dt">title =</span> <span class="st">&quot;boxplots&quot;</span>,
    <span class="dt">legend_gp =</span> <span class="kw">gpar</span>(<span class="dt">fill =</span> <span class="kw">c</span>(<span class="st">&quot;red&quot;</span>, <span class="st">&quot;green&quot;</span>)))
<span class="kw">draw</span>(ht_list, <span class="dt">padding =</span> <span class="kw">unit</span>(<span class="kw">c</span>(<span class="dv">20</span>, <span class="dv">2</span>, <span class="dv">2</span>, <span class="dv">2</span>), <span class="st">&quot;mm&quot;</span>), <span class="dt">heatmap_legend_list =</span> <span class="kw">list</span>(lgd))</code></pre></div>
<p><img src="09-examples_files/figure-html/unnamed-chunk-28-1.png" width="480" style="display: block; margin: auto;" /></p>

</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="upset-plot.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="other-high-level-plots.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"google": false,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "google", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/jokergoo/ComplexHeatmap-reference/edit/master/09-examples.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"download": ["ComplexHeatmap-reference.epub"],
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
